// @react-grab/cli/dist/cli.cjs
#!/usr/bin/env node
'use strict';

var commander = require('commander');
var pc = require('picocolors');
var basePrompts = require('prompts');
var child_process = require('child_process');
var fs = require('fs');
var path = require('path');
var ni = require('@antfu/ni');
var ignore = require('ignore');
var os = require('os');
var process2 = require('process');
var jsonc = require('jsonc-parser');
var TOML = require('smol-toml');
var ora = require('ora');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var pc__default = /*#__PURE__*/_interopDefault(pc);
var basePrompts__default = /*#__PURE__*/_interopDefault(basePrompts);
var fs__default = /*#__PURE__*/_interopDefault(fs);
var path__default = /*#__PURE__*/_interopDefault(path);
var ignore__default = /*#__PURE__*/_interopDefault(ignore);
var os__default = /*#__PURE__*/_interopDefault(os);
var process2__default = /*#__PURE__*/_interopDefault(process2);
var jsonc__namespace = /*#__PURE__*/_interopNamespace(jsonc);
var TOML__namespace = /*#__PURE__*/_interopNamespace(TOML);
var ora__default = /*#__PURE__*/_interopDefault(ora);

// src/utils/is-non-interactive.ts
var AGENT_ENVIRONMENT_VARIABLES = [
  "CI",
  "CLAUDECODE",
  "CURSOR_AGENT",
  "CODEX_CI",
  "OPENCODE",
  "AMP_HOME",
  "AMI"
];
var isEnvironmentVariableSet = (variable) => Boolean(process.env[variable]);
var detectNonInteractive = (yesFlag) => yesFlag || AGENT_ENVIRONMENT_VARIABLES.some(isEnvironmentVariableSet) || !process.stdin.isTTY;
var highlighter = {
  error: pc__default.default.red,
  warn: pc__default.default.yellow,
  info: pc__default.default.cyan,
  success: pc__default.default.green,
  dim: pc__default.default.dim
};

// src/utils/logger.ts
var logger = {
  error(...args) {
    console.log(highlighter.error(args.join(" ")));
  },
  warn(...args) {
    console.log(highlighter.warn(args.join(" ")));
  },
  info(...args) {
    console.log(highlighter.info(args.join(" ")));
  },
  success(...args) {
    console.log(highlighter.success(args.join(" ")));
  },
  dim(...args) {
    console.log(highlighter.dim(args.join(" ")));
  },
  log(...args) {
    console.log(args.join(" "));
  },
  break() {
    console.log("");
  }
};

// src/utils/prompts.ts
var onCancel = () => {
  logger.break();
  logger.log("Cancelled.");
  logger.break();
  process.exit(0);
};
var prompts = (questions) => {
  return basePrompts__default.default(questions, { onCancel });
};
var VALID_PACKAGE_MANAGERS = /* @__PURE__ */ new Set([
  "npm",
  "yarn",
  "pnpm",
  "bun"
]);
var detectPackageManager = async (projectRoot) => {
  const detected = await ni.detect({ cwd: projectRoot });
  if (detected) {
    const managerName = detected.split("@")[0];
    if (VALID_PACKAGE_MANAGERS.has(managerName)) {
      return managerName;
    }
  }
  return "npm";
};
var detectFramework = (projectRoot) => {
  const packageJsonPath = path.join(projectRoot, "package.json");
  if (!fs.existsSync(packageJsonPath)) {
    return "unknown";
  }
  try {
    const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, "utf-8"));
    const allDependencies = {
      ...packageJson.dependencies,
      ...packageJson.devDependencies
    };
    if (allDependencies["next"]) {
      return "next";
    }
    if (allDependencies["@tanstack/react-start"]) {
      return "tanstack";
    }
    if (allDependencies["vite"]) {
      return "vite";
    }
    if (allDependencies["webpack"]) {
      return "webpack";
    }
    return "unknown";
  } catch {
    return "unknown";
  }
};
var detectNextRouterType = (projectRoot) => {
  const hasAppDir = fs.existsSync(path.join(projectRoot, "app"));
  const hasSrcAppDir = fs.existsSync(path.join(projectRoot, "src", "app"));
  const hasPagesDir = fs.existsSync(path.join(projectRoot, "pages"));
  const hasSrcPagesDir = fs.existsSync(path.join(projectRoot, "src", "pages"));
  if (hasAppDir || hasSrcAppDir) {
    return "app";
  }
  if (hasPagesDir || hasSrcPagesDir) {
    return "pages";
  }
  return "unknown";
};
var detectMonorepo = (projectRoot) => {
  if (fs.existsSync(path.join(projectRoot, "pnpm-workspace.yaml"))) {
    return true;
  }
  if (fs.existsSync(path.join(projectRoot, "lerna.json"))) {
    return true;
  }
  const packageJsonPath = path.join(projectRoot, "package.json");
  if (fs.existsSync(packageJsonPath)) {
    try {
      const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, "utf-8"));
      if (packageJson.workspaces) {
        return true;
      }
    } catch {
      return false;
    }
  }
  return false;
};
var getWorkspacePatterns = (projectRoot) => {
  const patterns = [];
  const pnpmWorkspacePath = path.join(projectRoot, "pnpm-workspace.yaml");
  if (fs.existsSync(pnpmWorkspacePath)) {
    const content = fs.readFileSync(pnpmWorkspacePath, "utf-8");
    const lines = content.split("\n");
    let inPackages = false;
    for (const line of lines) {
      if (line.match(/^packages:\s*$/)) {
        inPackages = true;
        continue;
      }
      if (inPackages) {
        if (line.match(/^[a-zA-Z]/) || line.trim() === "") {
          if (line.match(/^[a-zA-Z]/)) inPackages = false;
          continue;
        }
        const match = line.match(/^\s*-\s*['"]?([^'"#\n]+?)['"]?\s*$/);
        if (match) {
          patterns.push(match[1].trim());
        }
      }
    }
  }
  const lernaJsonPath = path.join(projectRoot, "lerna.json");
  if (fs.existsSync(lernaJsonPath)) {
    try {
      const lernaJson = JSON.parse(fs.readFileSync(lernaJsonPath, "utf-8"));
      if (Array.isArray(lernaJson.packages)) {
        patterns.push(...lernaJson.packages);
      }
    } catch {
    }
  }
  const packageJsonPath = path.join(projectRoot, "package.json");
  if (fs.existsSync(packageJsonPath)) {
    try {
      const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, "utf-8"));
      if (Array.isArray(packageJson.workspaces)) {
        patterns.push(...packageJson.workspaces);
      } else if (packageJson.workspaces?.packages) {
        patterns.push(...packageJson.workspaces.packages);
      }
    } catch {
    }
  }
  return [...new Set(patterns)];
};
var expandWorkspacePattern = (projectRoot, pattern) => {
  const isGlob = pattern.endsWith("/*");
  const cleanPattern = pattern.replace(/\/\*$/, "");
  const basePath = path.join(projectRoot, cleanPattern);
  if (!fs.existsSync(basePath)) return [];
  if (!isGlob) {
    const hasPackageJson = fs.existsSync(path.join(basePath, "package.json"));
    return hasPackageJson ? [basePath] : [];
  }
  const results = [];
  try {
    const entries = fs.readdirSync(basePath, { withFileTypes: true });
    for (const entry of entries) {
      if (!entry.isDirectory()) continue;
      const packageJsonPath = path.join(basePath, entry.name, "package.json");
      if (fs.existsSync(packageJsonPath)) {
        results.push(path.join(basePath, entry.name));
      }
    }
  } catch {
    return results;
  }
  return results;
};
var hasReactDependency = (projectPath) => {
  const packageJsonPath = path.join(projectPath, "package.json");
  if (!fs.existsSync(packageJsonPath)) return false;
  try {
    const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, "utf-8"));
    const allDeps = {
      ...packageJson.dependencies,
      ...packageJson.devDependencies
    };
    return Boolean(allDeps["react"] || allDeps["react-dom"]);
  } catch {
    return false;
  }
};
var buildReactProject = (projectPath) => {
  const framework = detectFramework(projectPath);
  const hasReact = hasReactDependency(projectPath);
  if (!hasReact && framework === "unknown") return null;
  let name = path.basename(projectPath);
  const packageJsonPath = path.join(projectPath, "package.json");
  try {
    const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, "utf-8"));
    name = packageJson.name || name;
  } catch {
  }
  return { name, path: projectPath, framework, hasReact };
};
var findWorkspaceProjects = (projectRoot) => {
  const patterns = getWorkspacePatterns(projectRoot);
  const projects = [];
  for (const pattern of patterns) {
    for (const projectPath of expandWorkspacePattern(projectRoot, pattern)) {
      const project = buildReactProject(projectPath);
      if (project) projects.push(project);
    }
  }
  return projects;
};
var ALWAYS_IGNORED_DIRECTORIES = [
  "node_modules",
  ".git",
  ".next",
  ".cache",
  ".turbo",
  "dist",
  "build",
  "coverage",
  "test-results"
];
var loadGitignore = (projectRoot) => {
  const ignorer = ignore__default.default().add(ALWAYS_IGNORED_DIRECTORIES);
  const gitignorePath = path.join(projectRoot, ".gitignore");
  if (fs.existsSync(gitignorePath)) {
    try {
      ignorer.add(fs.readFileSync(gitignorePath, "utf-8"));
    } catch {
    }
  }
  return ignorer;
};
var scanDirectoryForProjects = (rootDirectory, ignorer, maxDepth, currentDepth = 0) => {
  if (currentDepth >= maxDepth) return [];
  if (!fs.existsSync(rootDirectory)) return [];
  const projects = [];
  try {
    const entries = fs.readdirSync(rootDirectory, { withFileTypes: true });
    for (const entry of entries) {
      if (!entry.isDirectory()) continue;
      if (ignorer.ignores(entry.name)) continue;
      const entryPath = path.join(rootDirectory, entry.name);
      const hasPackageJson = fs.existsSync(path.join(entryPath, "package.json"));
      if (hasPackageJson) {
        const project = buildReactProject(entryPath);
        if (project) {
          projects.push(project);
          continue;
        }
      }
      projects.push(
        ...scanDirectoryForProjects(
          entryPath,
          ignorer,
          maxDepth,
          currentDepth + 1
        )
      );
    }
  } catch {
    return projects;
  }
  return projects;
};
var MAX_SCAN_DEPTH = 2;
var findReactProjects = (projectRoot) => {
  if (detectMonorepo(projectRoot)) {
    return findWorkspaceProjects(projectRoot);
  }
  const ignorer = loadGitignore(projectRoot);
  return scanDirectoryForProjects(projectRoot, ignorer, MAX_SCAN_DEPTH);
};
var hasReactGrabInFile = (filePath) => {
  if (!fs.existsSync(filePath)) return false;
  try {
    const content = fs.readFileSync(filePath, "utf-8");
    const fuzzyPatterns = [
      /["'`][^"'`]*react-grab/,
      /react-grab[^"'`]*["'`]/,
      /<[^>]*react-grab/i,
      /import[^;]*react-grab/i,
      /require[^)]*react-grab/i,
      /from\s+[^;]*react-grab/i,
      /src[^>]*react-grab/i
    ];
    return fuzzyPatterns.some((pattern) => pattern.test(content));
  } catch {
    return false;
  }
};
var detectReactGrab = (projectRoot) => {
  const packageJsonPath = path.join(projectRoot, "package.json");
  if (fs.existsSync(packageJsonPath)) {
    try {
      const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, "utf-8"));
      const allDependencies = {
        ...packageJson.dependencies,
        ...packageJson.devDependencies
      };
      if (allDependencies["react-grab"]) {
        return true;
      }
    } catch {
    }
  }
  const filesToCheck = [
    path.join(projectRoot, "app", "layout.tsx"),
    path.join(projectRoot, "app", "layout.jsx"),
    path.join(projectRoot, "src", "app", "layout.tsx"),
    path.join(projectRoot, "src", "app", "layout.jsx"),
    path.join(projectRoot, "pages", "_document.tsx"),
    path.join(projectRoot, "pages", "_document.jsx"),
    path.join(projectRoot, "instrumentation-client.ts"),
    path.join(projectRoot, "instrumentation-client.js"),
    path.join(projectRoot, "src", "instrumentation-client.ts"),
    path.join(projectRoot, "src", "instrumentation-client.js"),
    path.join(projectRoot, "index.html"),
    path.join(projectRoot, "public", "index.html"),
    path.join(projectRoot, "src", "index.tsx"),
    path.join(projectRoot, "src", "index.ts"),
    path.join(projectRoot, "src", "main.tsx"),
    path.join(projectRoot, "src", "main.ts"),
    path.join(projectRoot, "src", "routes", "__root.tsx"),
    path.join(projectRoot, "src", "routes", "__root.jsx"),
    path.join(projectRoot, "app", "routes", "__root.tsx"),
    path.join(projectRoot, "app", "routes", "__root.jsx")
  ];
  return filesToCheck.some(hasReactGrabInFile);
};
var AGENT_PACKAGES = [
  "@react-grab/claude-code",
  "@react-grab/cursor",
  "@react-grab/opencode",
  "@react-grab/codex",
  "@react-grab/gemini",
  "@react-grab/amp",
  "@react-grab/ami",
  "@react-grab/droid",
  "@react-grab/copilot",
  "@react-grab/mcp"
];
var detectUnsupportedFramework = (projectRoot) => {
  const packageJsonPath = path.join(projectRoot, "package.json");
  if (!fs.existsSync(packageJsonPath)) {
    return null;
  }
  try {
    const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, "utf-8"));
    const allDependencies = {
      ...packageJson.dependencies,
      ...packageJson.devDependencies
    };
    if (allDependencies["@remix-run/react"] || allDependencies["remix"]) {
      return "remix";
    }
    if (allDependencies["astro"]) {
      return "astro";
    }
    if (allDependencies["@sveltejs/kit"]) {
      return "sveltekit";
    }
    if (allDependencies["gatsby"]) {
      return "gatsby";
    }
    return null;
  } catch {
    return null;
  }
};
var detectInstalledAgents = (projectRoot) => {
  const packageJsonPath = path.join(projectRoot, "package.json");
  if (!fs.existsSync(packageJsonPath)) {
    return [];
  }
  try {
    const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, "utf-8"));
    const allDependencies = {
      ...packageJson.dependencies,
      ...packageJson.devDependencies
    };
    return AGENT_PACKAGES.filter(
      (agent) => Boolean(allDependencies[agent])
    ).map((agent) => agent.replace("@react-grab/", ""));
  } catch {
    return [];
  }
};
var detectProject = async (projectRoot = process.cwd()) => {
  const framework = detectFramework(projectRoot);
  const packageManager = await detectPackageManager(projectRoot);
  return {
    packageManager,
    framework,
    nextRouterType: framework === "next" ? detectNextRouterType(projectRoot) : "unknown",
    isMonorepo: detectMonorepo(projectRoot),
    projectRoot,
    hasReactGrab: detectReactGrab(projectRoot),
    installedAgents: detectInstalledAgents(projectRoot),
    unsupportedFramework: detectUnsupportedFramework(projectRoot)
  };
};

// src/utils/diff.ts
var RED = "\x1B[31m";
var GREEN = "\x1B[32m";
var GRAY = "\x1B[90m";
var RESET = "\x1B[0m";
var BOLD = "\x1B[1m";
var generateDiff = (originalContent, newContent) => {
  const originalLines = originalContent.split("\n");
  const newLines = newContent.split("\n");
  const diff = [];
  Math.max(originalLines.length, newLines.length);
  let originalIndex = 0;
  let newIndex = 0;
  while (originalIndex < originalLines.length || newIndex < newLines.length) {
    const originalLine = originalLines[originalIndex];
    const newLine = newLines[newIndex];
    if (originalLine === newLine) {
      diff.push({
        type: "unchanged",
        content: originalLine,
        lineNumber: newIndex + 1
      });
      originalIndex++;
      newIndex++;
    } else if (originalLine === void 0) {
      diff.push({ type: "added", content: newLine, lineNumber: newIndex + 1 });
      newIndex++;
    } else if (newLine === void 0) {
      diff.push({ type: "removed", content: originalLine });
      originalIndex++;
    } else {
      const originalInNew = newLines.indexOf(originalLine, newIndex);
      const newInOriginal = originalLines.indexOf(newLine, originalIndex);
      if (originalInNew !== -1 && (newInOriginal === -1 || originalInNew - newIndex < newInOriginal - originalIndex)) {
        while (newIndex < originalInNew) {
          diff.push({
            type: "added",
            content: newLines[newIndex],
            lineNumber: newIndex + 1
          });
          newIndex++;
        }
      } else if (newInOriginal !== -1) {
        while (originalIndex < newInOriginal) {
          diff.push({ type: "removed", content: originalLines[originalIndex] });
          originalIndex++;
        }
      } else {
        diff.push({ type: "removed", content: originalLine });
        diff.push({
          type: "added",
          content: newLine,
          lineNumber: newIndex + 1
        });
        originalIndex++;
        newIndex++;
      }
    }
  }
  return diff;
};
var formatDiff = (diff, contextLines = 3) => {
  const lines = [];
  let lastPrintedIndex = -1;
  let hasChanges = false;
  const changedIndices = diff.map((line, index) => line.type !== "unchanged" ? index : -1).filter((index) => index !== -1);
  if (changedIndices.length === 0) {
    return `${GRAY}No changes${RESET}`;
  }
  for (const changedIndex of changedIndices) {
    const startContext = Math.max(0, changedIndex - contextLines);
    const endContext = Math.min(diff.length - 1, changedIndex + contextLines);
    if (startContext > lastPrintedIndex + 1 && lastPrintedIndex !== -1) {
      lines.push(`${GRAY}  ...${RESET}`);
    }
    for (let lineIndex = Math.max(startContext, lastPrintedIndex + 1); lineIndex <= endContext; lineIndex++) {
      const diffLine = diff[lineIndex];
      if (diffLine.type === "added") {
        lines.push(`${GREEN}+ ${diffLine.content}${RESET}`);
        hasChanges = true;
      } else if (diffLine.type === "removed") {
        lines.push(`${RED}- ${diffLine.content}${RESET}`);
        hasChanges = true;
      } else {
        lines.push(`${GRAY}  ${diffLine.content}${RESET}`);
      }
      lastPrintedIndex = lineIndex;
    }
  }
  return hasChanges ? lines.join("\n") : `${GRAY}No changes${RESET}`;
};
var printDiff = (filePath, originalContent, newContent) => {
  console.log(`
${BOLD}File: ${filePath}${RESET}`);
  console.log("\u2500".repeat(60));
  const diff = generateDiff(originalContent, newContent);
  console.log(formatDiff(diff));
  console.log("\u2500".repeat(60));
};

// src/utils/handle-error.ts
var handleError = (error) => {
  logger.break();
  logger.error(
    "Something went wrong. Please check the error below for more details."
  );
  logger.error("If the problem persists, please open an issue on GitHub.");
  logger.error("");
  if (error instanceof Error) {
    logger.error(error.message);
  }
  logger.break();
  process.exit(1);
};
var INSTALL_COMMANDS = {
  npm: "npm install",
  yarn: "yarn add",
  pnpm: "pnpm add",
  bun: "bun add"
};
var UNINSTALL_COMMANDS = {
  npm: "npm uninstall",
  yarn: "yarn remove",
  pnpm: "pnpm remove",
  bun: "bun remove"
};
var installPackages = (packages, packageManager, projectRoot, isDev = true) => {
  if (packages.length === 0) {
    return;
  }
  const command = INSTALL_COMMANDS[packageManager];
  const devFlag = isDev ? " -D" : "";
  const fullCommand = `${command}${devFlag} ${packages.join(" ")}`;
  console.log(`Running: ${fullCommand}
`);
  child_process.execSync(fullCommand, {
    cwd: projectRoot,
    stdio: "inherit"
  });
};
var getPackagesToInstall = (agent, includeReactGrab = true) => {
  const packages = [];
  if (includeReactGrab) {
    packages.push("react-grab");
  }
  if (agent !== "none") {
    packages.push(`@react-grab/${agent}`);
  }
  return packages;
};
var uninstallPackages = (packages, packageManager, projectRoot) => {
  if (packages.length === 0) {
    return;
  }
  const command = UNINSTALL_COMMANDS[packageManager];
  const fullCommand = `${command} ${packages.join(" ")}`;
  console.log(`Running: ${fullCommand}
`);
  child_process.execSync(fullCommand, {
    cwd: projectRoot,
    stdio: "inherit"
  });
};
var getPackagesToUninstall = (agent) => {
  return [`@react-grab/${agent}`];
};
var spinner = (text, options) => ora__default.default({ text, isSilent: options?.silent });

// src/utils/install-mcp.ts
var SERVER_NAME = "react-grab-mcp";
var PACKAGE_NAME = "@react-grab/mcp";
var getXdgConfigHome = () => process2__default.default.env.XDG_CONFIG_HOME || path__default.default.join(os__default.default.homedir(), ".config");
var getBaseDir = () => {
  const homeDir = os__default.default.homedir();
  if (process2__default.default.platform === "win32") {
    return process2__default.default.env.APPDATA || path__default.default.join(homeDir, "AppData", "Roaming");
  }
  if (process2__default.default.platform === "darwin") {
    return path__default.default.join(homeDir, "Library", "Application Support");
  }
  return getXdgConfigHome();
};
var getZedConfigPath = () => {
  if (process2__default.default.platform === "win32") {
    return path__default.default.join(getBaseDir(), "Zed", "settings.json");
  }
  return path__default.default.join(os__default.default.homedir(), ".config", "zed", "settings.json");
};
var getOpenCodeConfigPath = () => {
  const configDir = path__default.default.join(getXdgConfigHome(), "opencode");
  const jsoncPath = path__default.default.join(configDir, "opencode.jsonc");
  const jsonPath = path__default.default.join(configDir, "opencode.json");
  if (fs__default.default.existsSync(jsoncPath)) return jsoncPath;
  if (fs__default.default.existsSync(jsonPath)) return jsonPath;
  return jsoncPath;
};
var getClients = () => {
  const homeDir = os__default.default.homedir();
  const baseDir = getBaseDir();
  const stdioConfig = {
    command: "npx",
    args: ["-y", PACKAGE_NAME, "--stdio"]
  };
  return [
    {
      name: "Claude Code",
      configPath: path__default.default.join(homeDir, ".claude.json"),
      configKey: "mcpServers",
      format: "json",
      serverConfig: stdioConfig
    },
    {
      name: "Codex",
      configPath: path__default.default.join(
        process2__default.default.env.CODEX_HOME || path__default.default.join(homeDir, ".codex"),
        "config.toml"
      ),
      configKey: "mcp_servers",
      format: "toml",
      serverConfig: stdioConfig
    },
    {
      name: "Cursor",
      configPath: path__default.default.join(homeDir, ".cursor", "mcp.json"),
      configKey: "mcpServers",
      format: "json",
      serverConfig: stdioConfig
    },
    {
      name: "OpenCode",
      configPath: getOpenCodeConfigPath(),
      configKey: "mcp",
      format: "json",
      serverConfig: {
        type: "local",
        command: ["npx", "-y", PACKAGE_NAME, "--stdio"]
      }
    },
    {
      name: "VS Code",
      configPath: path__default.default.join(baseDir, "Code", "User", "mcp.json"),
      configKey: "servers",
      format: "json",
      serverConfig: { type: "stdio", ...stdioConfig }
    },
    {
      name: "Amp",
      configPath: path__default.default.join(homeDir, ".config", "amp", "settings.json"),
      configKey: "amp.mcpServers",
      format: "json",
      serverConfig: stdioConfig
    },
    {
      name: "Droid",
      configPath: path__default.default.join(homeDir, ".factory", "mcp.json"),
      configKey: "mcpServers",
      format: "json",
      serverConfig: { type: "stdio", ...stdioConfig }
    },
    {
      name: "Windsurf",
      configPath: path__default.default.join(homeDir, ".codeium", "windsurf", "mcp_config.json"),
      configKey: "mcpServers",
      format: "json",
      serverConfig: stdioConfig
    },
    {
      name: "Zed",
      configPath: getZedConfigPath(),
      configKey: "context_servers",
      format: "json",
      serverConfig: { source: "custom", ...stdioConfig, env: {} }
    }
  ];
};
var ensureDirectory = (filePath) => {
  const directory = path__default.default.dirname(filePath);
  if (!fs__default.default.existsSync(directory)) {
    fs__default.default.mkdirSync(directory, { recursive: true });
  }
};
var JSONC_FORMAT_OPTIONS = {
  tabSize: 2,
  insertSpaces: true
};
var upsertIntoJsonc = (filePath, content, configKey, serverName, serverConfig) => {
  const edits = jsonc__namespace.modify(content, [configKey, serverName], serverConfig, {
    formattingOptions: JSONC_FORMAT_OPTIONS
  });
  fs__default.default.writeFileSync(filePath, jsonc__namespace.applyEdits(content, edits));
};
var installJsonClient = (client) => {
  ensureDirectory(client.configPath);
  const content = fs__default.default.existsSync(client.configPath) ? fs__default.default.readFileSync(client.configPath, "utf8") : "{}";
  upsertIntoJsonc(
    client.configPath,
    content,
    client.configKey,
    SERVER_NAME,
    client.serverConfig
  );
};
var installTomlClient = (client) => {
  ensureDirectory(client.configPath);
  const existingConfig = fs__default.default.existsSync(
    client.configPath
  ) ? TOML__namespace.parse(fs__default.default.readFileSync(client.configPath, "utf8")) : {};
  const serverSection = existingConfig[client.configKey] ?? {};
  serverSection[SERVER_NAME] = client.serverConfig;
  existingConfig[client.configKey] = serverSection;
  fs__default.default.writeFileSync(client.configPath, TOML__namespace.stringify(existingConfig));
};
var getMcpClientNames = () => getClients().map((client) => client.name);
var installMcpServers = (selectedClients) => {
  const allClients = getClients();
  const clients = selectedClients ? allClients.filter((client) => selectedClients.includes(client.name)) : allClients;
  const results = [];
  const installSpinner = spinner("Installing MCP server.").start();
  for (const client of clients) {
    try {
      if (client.format === "toml") {
        installTomlClient(client);
      } else {
        installJsonClient(client);
      }
      results.push({
        client: client.name,
        configPath: client.configPath,
        success: true
      });
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      results.push({
        client: client.name,
        configPath: client.configPath,
        success: false,
        error: message
      });
    }
  }
  const successCount = results.filter((result) => result.success).length;
  if (successCount < results.length) {
    installSpinner.warn(
      `Installed to ${successCount}/${results.length} agents.`
    );
  } else {
    installSpinner.succeed(`Installed to ${successCount} agents.`);
  }
  for (const result of results) {
    if (result.success) {
      logger.log(
        `  ${highlighter.success("\u2713")} ${result.client} ${highlighter.dim("\u2192")} ${highlighter.dim(result.configPath)}`
      );
    } else {
      logger.log(
        `  ${highlighter.error("\u2717")} ${result.client} ${highlighter.dim("\u2192")} ${result.error}`
      );
    }
  }
  return results;
};
var promptConnectionMode = async () => {
  const { connectionMode } = await prompts({
    type: "select",
    name: "connectionMode",
    message: "How would you like to connect?",
    choices: [
      {
        title: `MCP ${highlighter.dim("(recommended)")}`,
        description: "Installs to all supported agents at once",
        value: "mcp"
      },
      {
        title: "Legacy",
        description: "Install a per-project agent package",
        value: "legacy"
      }
    ]
  });
  return connectionMode;
};
var promptMcpInstall = async () => {
  const clientNames = getMcpClientNames();
  const { selectedAgents } = await prompts({
    type: "multiselect",
    name: "selectedAgents",
    message: "Select agents to install MCP server for:",
    choices: clientNames.map((name) => ({
      title: name,
      value: name,
      selected: true
    }))
  });
  if (selectedAgents === void 0 || selectedAgents.length === 0) {
    return false;
  }
  logger.break();
  const results = installMcpServers(selectedAgents);
  const hasSuccess = results.some((result) => result.success);
  return hasSuccess;
};

// src/utils/templates.ts
var AGENTS = [
  "claude-code",
  "cursor",
  "opencode",
  "codex",
  "gemini",
  "amp",
  "ami",
  "droid",
  "copilot"
];
var AGENT_NAMES = {
  "claude-code": "Claude Code",
  cursor: "Cursor",
  opencode: "OpenCode",
  codex: "Codex",
  gemini: "Gemini",
  amp: "Amp",
  ami: "Ami",
  droid: "Droid",
  copilot: "Copilot"
};
var getAgentDisplayName = (agent) => {
  if (agent === "mcp") return "MCP";
  if (agent in AGENT_NAMES) {
    return AGENT_NAMES[agent];
  }
  return agent;
};
var NEXT_APP_ROUTER_SCRIPT = `{process.env.NODE_ENV === "development" && (
          <Script
            src="//unpkg.com/react-grab/dist/index.global.js"
            crossOrigin="anonymous"
            strategy="beforeInteractive"
          />
        )}`;
var NEXT_APP_ROUTER_SCRIPT_WITH_AGENT = (agent) => {
  if (agent === "none") return NEXT_APP_ROUTER_SCRIPT;
  return `{process.env.NODE_ENV === "development" && (
          <Script
            src="//unpkg.com/react-grab/dist/index.global.js"
            crossOrigin="anonymous"
            strategy="beforeInteractive"
          />
        )}
        {process.env.NODE_ENV === "development" && (
          <Script
            src="//unpkg.com/@react-grab/${agent}/dist/client.global.js"
            strategy="lazyOnload"
          />
        )}`;
};
var NEXT_PAGES_ROUTER_SCRIPT = `{process.env.NODE_ENV === "development" && (
          <Script
            src="//unpkg.com/react-grab/dist/index.global.js"
            crossOrigin="anonymous"
            strategy="beforeInteractive"
          />
        )}`;
var NEXT_PAGES_ROUTER_SCRIPT_WITH_AGENT = (agent) => {
  if (agent === "none") return NEXT_PAGES_ROUTER_SCRIPT;
  return `{process.env.NODE_ENV === "development" && (
          <Script
            src="//unpkg.com/react-grab/dist/index.global.js"
            crossOrigin="anonymous"
            strategy="beforeInteractive"
          />
        )}
        {process.env.NODE_ENV === "development" && (
          <Script
            src="//unpkg.com/@react-grab/${agent}/dist/client.global.js"
            strategy="lazyOnload"
          />
        )}`;
};
var VITE_SCRIPT = `<script type="module">
      if (import.meta.env.DEV) {
        import("react-grab");
      }
    </script>`;
var VITE_SCRIPT_WITH_AGENT = (agent) => {
  if (agent === "none") return VITE_SCRIPT;
  return `<script type="module">
      if (import.meta.env.DEV) {
        import("react-grab");
        import("@react-grab/${agent}/client");
      }
    </script>`;
};
var WEBPACK_IMPORT = `if (process.env.NODE_ENV === "development") {
  import("react-grab");
}`;
var WEBPACK_IMPORT_WITH_AGENT = (agent) => {
  if (agent === "none") return WEBPACK_IMPORT;
  return `if (process.env.NODE_ENV === "development") {
  import("react-grab");
  import("@react-grab/${agent}/client");
}`;
};
var TANSTACK_EFFECT = `useEffect(() => {
    if (import.meta.env.DEV) {
      void import("react-grab");
    }
  }, []);`;
var TANSTACK_EFFECT_WITH_AGENT = (agent) => {
  if (agent === "none") return TANSTACK_EFFECT;
  return `useEffect(() => {
    if (import.meta.env.DEV) {
      void import("react-grab");
      void import("@react-grab/${agent}/client");
    }
  }, []);`;
};
var SCRIPT_IMPORT = 'import Script from "next/script";';
var hasReactGrabCode = (content) => {
  const fuzzyPatterns = [
    /["'`][^"'`]*react-grab/,
    /react-grab[^"'`]*["'`]/,
    /<[^>]*react-grab/i,
    /import[^;]*react-grab/i,
    /require[^)]*react-grab/i,
    /from\s+[^;]*react-grab/i,
    /src[^>]*react-grab/i,
    /href[^>]*react-grab/i
  ];
  return fuzzyPatterns.some((pattern) => pattern.test(content));
};
var findLayoutFile = (projectRoot) => {
  const possiblePaths = [
    path.join(projectRoot, "app", "layout.tsx"),
    path.join(projectRoot, "app", "layout.jsx"),
    path.join(projectRoot, "src", "app", "layout.tsx"),
    path.join(projectRoot, "src", "app", "layout.jsx")
  ];
  for (const filePath of possiblePaths) {
    if (fs.existsSync(filePath)) {
      return filePath;
    }
  }
  return null;
};
var findInstrumentationFile = (projectRoot) => {
  const possiblePaths = [
    path.join(projectRoot, "instrumentation-client.ts"),
    path.join(projectRoot, "instrumentation-client.js"),
    path.join(projectRoot, "src", "instrumentation-client.ts"),
    path.join(projectRoot, "src", "instrumentation-client.js")
  ];
  for (const filePath of possiblePaths) {
    if (fs.existsSync(filePath)) {
      return filePath;
    }
  }
  return null;
};
var hasReactGrabInInstrumentation = (projectRoot) => {
  const instrumentationPath = findInstrumentationFile(projectRoot);
  if (!instrumentationPath) return false;
  const content = fs.readFileSync(instrumentationPath, "utf-8");
  return hasReactGrabCode(content);
};
var findDocumentFile = (projectRoot) => {
  const possiblePaths = [
    path.join(projectRoot, "pages", "_document.tsx"),
    path.join(projectRoot, "pages", "_document.jsx"),
    path.join(projectRoot, "src", "pages", "_document.tsx"),
    path.join(projectRoot, "src", "pages", "_document.jsx")
  ];
  for (const filePath of possiblePaths) {
    if (fs.existsSync(filePath)) {
      return filePath;
    }
  }
  return null;
};
var findIndexHtml = (projectRoot) => {
  const possiblePaths = [
    path.join(projectRoot, "index.html"),
    path.join(projectRoot, "public", "index.html")
  ];
  for (const filePath of possiblePaths) {
    if (fs.existsSync(filePath)) {
      return filePath;
    }
  }
  return null;
};
var findEntryFile = (projectRoot) => {
  const possiblePaths = [
    path.join(projectRoot, "src", "index.tsx"),
    path.join(projectRoot, "src", "index.jsx"),
    path.join(projectRoot, "src", "index.ts"),
    path.join(projectRoot, "src", "index.js"),
    path.join(projectRoot, "src", "main.tsx"),
    path.join(projectRoot, "src", "main.jsx"),
    path.join(projectRoot, "src", "main.ts"),
    path.join(projectRoot, "src", "main.js")
  ];
  for (const filePath of possiblePaths) {
    if (fs.existsSync(filePath)) {
      return filePath;
    }
  }
  return null;
};
var findTanStackRootFile = (projectRoot) => {
  const possiblePaths = [
    path.join(projectRoot, "src", "routes", "__root.tsx"),
    path.join(projectRoot, "src", "routes", "__root.jsx"),
    path.join(projectRoot, "app", "routes", "__root.tsx"),
    path.join(projectRoot, "app", "routes", "__root.jsx")
  ];
  for (const filePath of possiblePaths) {
    if (fs.existsSync(filePath)) {
      return filePath;
    }
  }
  return null;
};
var addAgentToExistingNextApp = (originalContent, agent, filePath) => {
  if (agent === "none") {
    return {
      success: true,
      filePath,
      message: "React Grab is already configured",
      noChanges: true
    };
  }
  const agentPackage = `@react-grab/${agent}`;
  if (originalContent.includes(agentPackage)) {
    return {
      success: true,
      filePath,
      message: `Agent ${agent} is already configured`,
      noChanges: true
    };
  }
  const agentScript = `{process.env.NODE_ENV === "development" && (
          <Script
            src="//unpkg.com/${agentPackage}/dist/client.global.js"
            strategy="lazyOnload"
          />
        )}`;
  const reactGrabBlockMatch = originalContent.match(
    /\{process\.env\.NODE_ENV\s*===\s*["']development["']\s*&&\s*\(\s*<Script[^>]*react-grab[^>]*\/>\s*\)\}/is
  );
  if (reactGrabBlockMatch) {
    const newContent = originalContent.replace(
      reactGrabBlockMatch[0],
      `${reactGrabBlockMatch[0]}
        ${agentScript}`
    );
    return {
      success: true,
      filePath,
      message: `Add ${agent} agent`,
      originalContent,
      newContent
    };
  }
  const bareScriptMatch = originalContent.match(
    /<Script[^>]*react-grab[^>]*\/>/i
  );
  if (bareScriptMatch) {
    const newContent = originalContent.replace(
      bareScriptMatch[0],
      `${bareScriptMatch[0]}
        <Script src="//unpkg.com/${agentPackage}/dist/client.global.js" strategy="lazyOnload" />`
    );
    return {
      success: true,
      filePath,
      message: `Add ${agent} agent`,
      originalContent,
      newContent
    };
  }
  return {
    success: false,
    filePath,
    message: "Could not find React Grab script to add agent after"
  };
};
var addAgentToExistingVite = (originalContent, agent, filePath) => {
  if (agent === "none") {
    return {
      success: true,
      filePath,
      message: "React Grab is already configured",
      noChanges: true
    };
  }
  const agentPackage = `@react-grab/${agent}`;
  if (originalContent.includes(agentPackage)) {
    return {
      success: true,
      filePath,
      message: `Agent ${agent} is already configured`,
      noChanges: true
    };
  }
  const agentImport = `import("${agentPackage}/client");`;
  const reactGrabImportMatch = originalContent.match(
    /import\s*\(\s*["']react-grab["']\s*\);?/
  );
  if (reactGrabImportMatch) {
    const matchedText = reactGrabImportMatch[0];
    const hasSemicolon = matchedText.endsWith(";");
    const newContent = originalContent.replace(
      matchedText,
      `${hasSemicolon ? matchedText.slice(0, -1) : matchedText};
        ${agentImport}`
    );
    return {
      success: true,
      filePath,
      message: `Add ${agent} agent`,
      originalContent,
      newContent
    };
  }
  return {
    success: false,
    filePath,
    message: "Could not find React Grab import to add agent after"
  };
};
var addAgentToExistingWebpack = (originalContent, agent, filePath) => {
  if (agent === "none") {
    return {
      success: true,
      filePath,
      message: "React Grab is already configured",
      noChanges: true
    };
  }
  const agentPackage = `@react-grab/${agent}`;
  if (originalContent.includes(agentPackage)) {
    return {
      success: true,
      filePath,
      message: `Agent ${agent} is already configured`,
      noChanges: true
    };
  }
  const agentImport = `import("${agentPackage}/client");`;
  const reactGrabImportMatch = originalContent.match(
    /import\s*\(\s*["']react-grab["']\s*\);?/
  );
  if (reactGrabImportMatch) {
    const matchedText = reactGrabImportMatch[0];
    const hasSemicolon = matchedText.endsWith(";");
    const newContent = originalContent.replace(
      matchedText,
      `${hasSemicolon ? matchedText.slice(0, -1) : matchedText};
  ${agentImport}`
    );
    return {
      success: true,
      filePath,
      message: `Add ${agent} agent`,
      originalContent,
      newContent
    };
  }
  return {
    success: false,
    filePath,
    message: "Could not find React Grab import to add agent after"
  };
};
var addAgentToExistingTanStack = (originalContent, agent, filePath) => {
  if (agent === "none") {
    return {
      success: true,
      filePath,
      message: "React Grab is already configured",
      noChanges: true
    };
  }
  const agentPackage = `@react-grab/${agent}`;
  if (originalContent.includes(agentPackage)) {
    return {
      success: true,
      filePath,
      message: `Agent ${agent} is already configured`,
      noChanges: true
    };
  }
  const agentImport = `void import("${agentPackage}/client");`;
  const reactGrabImportMatch = originalContent.match(
    /void\s+import\s*\(\s*["']react-grab["']\s*\);?/
  );
  if (reactGrabImportMatch) {
    const matchedText = reactGrabImportMatch[0];
    const hasSemicolon = matchedText.endsWith(";");
    const newContent = originalContent.replace(
      matchedText,
      `${hasSemicolon ? matchedText.slice(0, -1) : matchedText};
      ${agentImport}`
    );
    return {
      success: true,
      filePath,
      message: `Add ${agent} agent`,
      originalContent,
      newContent
    };
  }
  return {
    success: false,
    filePath,
    message: "Could not find React Grab import to add agent after"
  };
};
var transformNextAppRouter = (projectRoot, agent, reactGrabAlreadyConfigured, force = false) => {
  const layoutPath = findLayoutFile(projectRoot);
  if (!layoutPath) {
    return {
      success: false,
      filePath: "",
      message: "Could not find app/layout.tsx or app/layout.jsx"
    };
  }
  const originalContent = fs.readFileSync(layoutPath, "utf-8");
  let newContent = originalContent;
  const hasReactGrabInFile2 = hasReactGrabCode(originalContent);
  const hasReactGrabInInstrumentationFile = hasReactGrabInInstrumentation(projectRoot);
  if (!force && hasReactGrabInFile2 && reactGrabAlreadyConfigured) {
    return addAgentToExistingNextApp(originalContent, agent, layoutPath);
  }
  if (!force && (hasReactGrabInFile2 || hasReactGrabInInstrumentationFile)) {
    return {
      success: true,
      filePath: layoutPath,
      message: "React Grab is already installed" + (hasReactGrabInInstrumentationFile ? " in instrumentation-client" : " in this file"),
      noChanges: true
    };
  }
  if (!newContent.includes('import Script from "next/script"')) {
    const importMatch = newContent.match(/^import .+ from ['"].+['"];?\s*$/m);
    if (importMatch) {
      newContent = newContent.replace(
        importMatch[0],
        `${importMatch[0]}
${SCRIPT_IMPORT}`
      );
    } else {
      newContent = `${SCRIPT_IMPORT}

${newContent}`;
    }
  }
  const scriptBlock = NEXT_APP_ROUTER_SCRIPT_WITH_AGENT(agent);
  const headMatch = newContent.match(/<head[^>]*>/);
  if (headMatch) {
    newContent = newContent.replace(
      headMatch[0],
      `${headMatch[0]}
        ${scriptBlock}`
    );
  } else {
    const htmlMatch = newContent.match(/<html[^>]*>/);
    if (htmlMatch) {
      newContent = newContent.replace(
        htmlMatch[0],
        `${htmlMatch[0]}
      <head>
        ${scriptBlock}
      </head>`
      );
    }
  }
  return {
    success: true,
    filePath: layoutPath,
    message: "Add React Grab" + (agent !== "none" ? ` with ${agent} agent` : ""),
    originalContent,
    newContent
  };
};
var transformNextPagesRouter = (projectRoot, agent, reactGrabAlreadyConfigured, force = false) => {
  const documentPath = findDocumentFile(projectRoot);
  if (!documentPath) {
    return {
      success: false,
      filePath: "",
      message: 'Could not find pages/_document.tsx or pages/_document.jsx.\n\nTo set up React Grab with Pages Router, create pages/_document.tsx with:\n\n  import { Html, Head, Main, NextScript } from "next/document";\n  import Script from "next/script";\n\n  export default function Document() {\n    return (\n      <Html>\n        <Head>\n          {process.env.NODE_ENV === "development" && (\n            <Script src="//unpkg.com/react-grab/dist/index.global.js" strategy="beforeInteractive" />\n          )}\n        </Head>\n        <body>\n          <Main />\n          <NextScript />\n        </body>\n      </Html>\n    );\n  }'
    };
  }
  const originalContent = fs.readFileSync(documentPath, "utf-8");
  let newContent = originalContent;
  const hasReactGrabInFile2 = hasReactGrabCode(originalContent);
  const hasReactGrabInInstrumentationFile = hasReactGrabInInstrumentation(projectRoot);
  if (!force && hasReactGrabInFile2 && reactGrabAlreadyConfigured) {
    return addAgentToExistingNextApp(originalContent, agent, documentPath);
  }
  if (!force && (hasReactGrabInFile2 || hasReactGrabInInstrumentationFile)) {
    return {
      success: true,
      filePath: documentPath,
      message: "React Grab is already installed" + (hasReactGrabInInstrumentationFile ? " in instrumentation-client" : " in this file"),
      noChanges: true
    };
  }
  if (!newContent.includes('import Script from "next/script"')) {
    const importMatch = newContent.match(/^import .+ from ['"].+['"];?\s*$/m);
    if (importMatch) {
      newContent = newContent.replace(
        importMatch[0],
        `${importMatch[0]}
${SCRIPT_IMPORT}`
      );
    }
  }
  const scriptBlock = NEXT_PAGES_ROUTER_SCRIPT_WITH_AGENT(agent);
  const headMatch = newContent.match(/<Head[^>]*>/);
  if (headMatch) {
    newContent = newContent.replace(
      headMatch[0],
      `${headMatch[0]}
        ${scriptBlock}`
    );
  }
  return {
    success: true,
    filePath: documentPath,
    message: "Add React Grab" + (agent !== "none" ? ` with ${agent} agent` : ""),
    originalContent,
    newContent
  };
};
var transformVite = (projectRoot, agent, reactGrabAlreadyConfigured, force = false) => {
  const indexPath = findIndexHtml(projectRoot);
  if (!indexPath) {
    return {
      success: false,
      filePath: "",
      message: "Could not find index.html"
    };
  }
  const originalContent = fs.readFileSync(indexPath, "utf-8");
  let newContent = originalContent;
  const hasReactGrabInFile2 = hasReactGrabCode(originalContent);
  if (!force && hasReactGrabInFile2 && reactGrabAlreadyConfigured) {
    return addAgentToExistingVite(originalContent, agent, indexPath);
  }
  if (!force && hasReactGrabInFile2) {
    return {
      success: true,
      filePath: indexPath,
      message: "React Grab is already installed in this file",
      noChanges: true
    };
  }
  const scriptBlock = VITE_SCRIPT_WITH_AGENT(agent);
  const headMatch = newContent.match(/<head[^>]*>/i);
  if (headMatch) {
    newContent = newContent.replace(
      headMatch[0],
      `${headMatch[0]}
    ${scriptBlock}`
    );
  }
  return {
    success: true,
    filePath: indexPath,
    message: "Add React Grab" + (agent !== "none" ? ` with ${agent} agent` : ""),
    originalContent,
    newContent
  };
};
var transformWebpack = (projectRoot, agent, reactGrabAlreadyConfigured, force = false) => {
  const entryPath = findEntryFile(projectRoot);
  if (!entryPath) {
    return {
      success: false,
      filePath: "",
      message: "Could not find entry file (src/index.tsx, src/main.tsx, etc.)"
    };
  }
  const originalContent = fs.readFileSync(entryPath, "utf-8");
  const hasReactGrabInFile2 = hasReactGrabCode(originalContent);
  if (!force && hasReactGrabInFile2 && reactGrabAlreadyConfigured) {
    return addAgentToExistingWebpack(originalContent, agent, entryPath);
  }
  if (!force && hasReactGrabInFile2) {
    return {
      success: true,
      filePath: entryPath,
      message: "React Grab is already installed in this file",
      noChanges: true
    };
  }
  const importBlock = WEBPACK_IMPORT_WITH_AGENT(agent);
  const newContent = `${importBlock}

${originalContent}`;
  return {
    success: true,
    filePath: entryPath,
    message: "Add React Grab" + (agent !== "none" ? ` with ${agent} agent` : ""),
    originalContent,
    newContent
  };
};
var transformTanStack = (projectRoot, agent, reactGrabAlreadyConfigured, force = false) => {
  const rootPath = findTanStackRootFile(projectRoot);
  if (!rootPath) {
    return {
      success: false,
      filePath: "",
      message: 'Could not find src/routes/__root.tsx or app/routes/__root.tsx.\n\nTo set up React Grab with TanStack Start, add this to your root route component:\n\n  import { useEffect } from "react";\n\n  useEffect(() => {\n    if (import.meta.env.DEV) {\n      void import("react-grab");\n    }\n  }, []);'
    };
  }
  const originalContent = fs.readFileSync(rootPath, "utf-8");
  let newContent = originalContent;
  const hasReactGrabInFile2 = hasReactGrabCode(originalContent);
  if (!force && hasReactGrabInFile2 && reactGrabAlreadyConfigured) {
    return addAgentToExistingTanStack(originalContent, agent, rootPath);
  }
  if (!force && hasReactGrabInFile2) {
    return {
      success: true,
      filePath: rootPath,
      message: "React Grab is already installed in this file",
      noChanges: true
    };
  }
  const hasUseEffectImport = /import\s+\{[^}]*useEffect[^}]*\}\s+from\s+["']react["']/.test(newContent);
  if (!hasUseEffectImport) {
    const reactImportMatch = newContent.match(
      /import\s+\{([^}]*)\}\s+from\s+["']react["'];?/
    );
    if (reactImportMatch) {
      const existingImports = reactImportMatch[1];
      newContent = newContent.replace(
        reactImportMatch[0],
        `import { ${existingImports.trim()}, useEffect } from "react";`
      );
    } else {
      const firstImportMatch = newContent.match(
        /^import .+ from ['"].+['"];?\s*$/m
      );
      if (firstImportMatch) {
        newContent = newContent.replace(
          firstImportMatch[0],
          `import { useEffect } from "react";
${firstImportMatch[0]}`
        );
      } else {
        newContent = `import { useEffect } from "react";

${newContent}`;
      }
    }
  }
  const effectBlock = TANSTACK_EFFECT_WITH_AGENT(agent);
  const componentMatch = newContent.match(/function\s+(\w+)\s*\([^)]*\)\s*\{/);
  if (componentMatch) {
    const insertPosition = componentMatch.index + componentMatch[0].length;
    newContent = newContent.slice(0, insertPosition) + `
  ${effectBlock}
` + newContent.slice(insertPosition);
  } else {
    return {
      success: false,
      filePath: rootPath,
      message: "Could not find a component function in the root file"
    };
  }
  return {
    success: true,
    filePath: rootPath,
    message: "Add React Grab" + (agent !== "none" ? ` with ${agent} agent` : ""),
    originalContent,
    newContent
  };
};
var previewTransform = (projectRoot, framework, nextRouterType, agent, reactGrabAlreadyConfigured = false, force = false) => {
  switch (framework) {
    case "next":
      if (nextRouterType === "app") {
        return transformNextAppRouter(
          projectRoot,
          agent,
          reactGrabAlreadyConfigured,
          force
        );
      }
      return transformNextPagesRouter(
        projectRoot,
        agent,
        reactGrabAlreadyConfigured,
        force
      );
    case "vite":
      return transformVite(
        projectRoot,
        agent,
        reactGrabAlreadyConfigured,
        force
      );
    case "tanstack":
      return transformTanStack(
        projectRoot,
        agent,
        reactGrabAlreadyConfigured,
        force
      );
    case "webpack":
      return transformWebpack(
        projectRoot,
        agent,
        reactGrabAlreadyConfigured,
        force
      );
    default:
      return {
        success: false,
        filePath: "",
        message: `Unknown framework: ${framework}. Please add React Grab manually.`
      };
  }
};
var canWriteToFile = (filePath) => {
  try {
    fs.accessSync(filePath, fs.constants.W_OK);
    return true;
  } catch {
    return false;
  }
};
var applyTransform = (result) => {
  if (result.success && result.newContent && result.filePath) {
    if (!canWriteToFile(result.filePath)) {
      return {
        success: false,
        error: `Cannot write to ${result.filePath}. Check file permissions.`
      };
    }
    try {
      fs.writeFileSync(result.filePath, result.newContent);
      return { success: true };
    } catch (error) {
      return {
        success: false,
        error: `Failed to write to ${result.filePath}: ${error instanceof Error ? error.message : "Unknown error"}`
      };
    }
  }
  return { success: true };
};
var getPackageExecutor = (packageManager) => {
  switch (packageManager) {
    case "bun":
      return "bunx";
    case "pnpm":
      return "pnpm dlx";
    case "yarn":
      return "npx";
    case "npm":
    default:
      return "npx";
  }
};
var AGENT_PACKAGES2 = {
  "claude-code": "@react-grab/claude-code@latest",
  cursor: "@react-grab/cursor@latest",
  opencode: "@react-grab/opencode@latest",
  codex: "@react-grab/codex@latest",
  gemini: "@react-grab/gemini@latest",
  amp: "@react-grab/amp@latest",
  droid: "@react-grab/droid@latest",
  copilot: "@react-grab/copilot@latest"
};
var getAgentPrefix = (agent, packageManager) => {
  const agentPackage = AGENT_PACKAGES2[agent];
  if (!agentPackage) return null;
  const executor = getPackageExecutor(packageManager);
  return `${executor} ${agentPackage} &&`;
};
var getAllAgentPrefixVariants = (agent) => {
  const agentPackage = AGENT_PACKAGES2[agent];
  if (!agentPackage) return [];
  return [
    `npx ${agentPackage} &&`,
    `bunx ${agentPackage} &&`,
    `pnpm dlx ${agentPackage} &&`,
    `yarn dlx ${agentPackage} &&`
  ];
};
var previewPackageJsonTransform = (projectRoot, agent, installedAgents, packageManager = "npm") => {
  if (agent === "none") {
    return {
      success: true,
      filePath: "",
      message: "No agent selected, skipping package.json modification",
      noChanges: true
    };
  }
  if (agent === "mcp") {
    return {
      success: true,
      filePath: "",
      message: "MCP does not use package.json dev script",
      noChanges: true
    };
  }
  const packageJsonPath = path.join(projectRoot, "package.json");
  if (!fs.existsSync(packageJsonPath)) {
    return {
      success: false,
      filePath: "",
      message: "Could not find package.json"
    };
  }
  const originalContent = fs.readFileSync(packageJsonPath, "utf-8");
  const agentPrefix = getAgentPrefix(agent, packageManager);
  if (!agentPrefix) {
    return {
      success: false,
      filePath: packageJsonPath,
      message: `Unknown agent: ${agent}`
    };
  }
  const allPrefixVariants = getAllAgentPrefixVariants(agent);
  const hasExistingPrefix = allPrefixVariants.some(
    (prefix) => originalContent.includes(prefix)
  );
  if (hasExistingPrefix) {
    return {
      success: true,
      filePath: packageJsonPath,
      message: `Agent ${agent} dev script is already configured`,
      noChanges: true
    };
  }
  try {
    const packageJson = JSON.parse(originalContent);
    let targetScriptKey = "dev";
    if (!packageJson.scripts?.dev) {
      const devScriptKeys = Object.keys(packageJson.scripts || {}).filter(
        (key) => key.startsWith("dev")
      );
      if (devScriptKeys.length > 0) {
        targetScriptKey = devScriptKeys[0];
      } else {
        return {
          success: true,
          filePath: packageJsonPath,
          message: "No dev script found in package.json",
          noChanges: true,
          warning: `Could not inject agent into package.json (no dev script found).
Run this command manually before starting your dev server:
  ${agentPrefix} <your dev command>`
        };
      }
    }
    const currentDevScript = packageJson.scripts[targetScriptKey];
    for (const installedAgent of installedAgents) {
      const installedPrefixVariants = getAllAgentPrefixVariants(installedAgent);
      const hasInstalledAgentPrefix = installedPrefixVariants.some(
        (prefix) => currentDevScript.includes(prefix)
      );
      if (hasInstalledAgentPrefix) {
        return {
          success: true,
          filePath: packageJsonPath,
          message: `Agent ${installedAgent} is already in ${targetScriptKey} script`,
          noChanges: true
        };
      }
    }
    packageJson.scripts[targetScriptKey] = `${agentPrefix} ${currentDevScript}`;
    const newContent = JSON.stringify(packageJson, null, 2) + "\n";
    return {
      success: true,
      filePath: packageJsonPath,
      message: `Add ${agent} server to ${targetScriptKey} script`,
      originalContent,
      newContent
    };
  } catch {
    return {
      success: false,
      filePath: packageJsonPath,
      message: "Failed to parse package.json"
    };
  }
};
var applyPackageJsonTransform = (result) => {
  if (result.success && result.newContent && result.filePath) {
    if (!canWriteToFile(result.filePath)) {
      return {
        success: false,
        error: `Cannot write to ${result.filePath}. Check file permissions.`
      };
    }
    try {
      fs.writeFileSync(result.filePath, result.newContent);
      return { success: true };
    } catch (error) {
      return {
        success: false,
        error: `Failed to write to ${result.filePath}: ${error instanceof Error ? error.message : "Unknown error"}`
      };
    }
  }
  return { success: true };
};
var formatOptionsForNextjs = (options) => {
  const parts = [];
  if (options.activationKey) {
    parts.push(`activationKey: ${JSON.stringify(options.activationKey)}`);
  }
  if (options.activationMode) {
    parts.push(`activationMode: "${options.activationMode}"`);
  }
  if (options.keyHoldDuration !== void 0) {
    parts.push(`keyHoldDuration: ${options.keyHoldDuration}`);
  }
  if (options.allowActivationInsideInput !== void 0) {
    parts.push(
      `allowActivationInsideInput: ${options.allowActivationInsideInput}`
    );
  }
  if (options.maxContextLines !== void 0) {
    parts.push(`maxContextLines: ${options.maxContextLines}`);
  }
  return `{ ${parts.join(", ")} }`;
};
var formatOptionsAsJson = (options) => {
  const cleanOptions = {};
  if (options.activationKey) {
    cleanOptions.activationKey = options.activationKey;
  }
  if (options.activationMode) {
    cleanOptions.activationMode = options.activationMode;
  }
  if (options.keyHoldDuration !== void 0) {
    cleanOptions.keyHoldDuration = options.keyHoldDuration;
  }
  if (options.allowActivationInsideInput !== void 0) {
    cleanOptions.allowActivationInsideInput = options.allowActivationInsideInput;
  }
  if (options.maxContextLines !== void 0) {
    cleanOptions.maxContextLines = options.maxContextLines;
  }
  return JSON.stringify(cleanOptions);
};
var findReactGrabFile = (projectRoot, framework, nextRouterType) => {
  switch (framework) {
    case "next":
      if (nextRouterType === "app") {
        return findLayoutFile(projectRoot);
      }
      return findDocumentFile(projectRoot);
    case "vite":
      return findIndexHtml(projectRoot);
    case "tanstack":
      return findTanStackRootFile(projectRoot);
    case "webpack":
      return findEntryFile(projectRoot);
    default:
      return null;
  }
};
var addOptionsToNextScript = (originalContent, options, filePath) => {
  const reactGrabScriptMatch = originalContent.match(
    /(<Script[\s\S]*?react-grab[\s\S]*?)\s*(\/?>)/i
  );
  if (!reactGrabScriptMatch) {
    return {
      success: false,
      filePath,
      message: "Could not find React Grab Script tag"
    };
  }
  const scriptTag = reactGrabScriptMatch[0];
  const scriptOpening = reactGrabScriptMatch[1];
  const scriptClosing = reactGrabScriptMatch[2];
  const existingDataOptionsMatch = scriptTag.match(
    /data-options=\{JSON\.stringify\([^)]+\)\}/
  );
  const dataOptionsAttr = `data-options={JSON.stringify(
              ${formatOptionsForNextjs(options)}
            )}`;
  let newScriptTag;
  if (existingDataOptionsMatch) {
    newScriptTag = scriptTag.replace(
      existingDataOptionsMatch[0],
      dataOptionsAttr
    );
  } else {
    newScriptTag = `${scriptOpening}
            ${dataOptionsAttr}
          ${scriptClosing}`;
  }
  const newContent = originalContent.replace(scriptTag, newScriptTag);
  return {
    success: true,
    filePath,
    message: "Update React Grab options",
    originalContent,
    newContent
  };
};
var addOptionsToViteScript = (originalContent, options, filePath) => {
  const reactGrabImportWithInitMatch = originalContent.match(
    /import\s*\(\s*["']react-grab["']\s*\)(?:\.then\s*\(\s*\(m\)\s*=>\s*m\.init\s*\([^)]*\)\s*\))?/
  );
  if (!reactGrabImportWithInitMatch) {
    return {
      success: false,
      filePath,
      message: "Could not find React Grab import"
    };
  }
  const optionsJson = formatOptionsAsJson(options);
  const newImport = `import("react-grab").then((m) => m.init(${optionsJson}))`;
  const newContent = originalContent.replace(
    reactGrabImportWithInitMatch[0],
    newImport
  );
  return {
    success: true,
    filePath,
    message: "Update React Grab options",
    originalContent,
    newContent
  };
};
var addOptionsToWebpackImport = (originalContent, options, filePath) => {
  const reactGrabImportWithInitMatch = originalContent.match(
    /import\s*\(\s*["']react-grab["']\s*\)(?:\.then\s*\(\s*\(m\)\s*=>\s*m\.init\s*\([^)]*\)\s*\))?/
  );
  if (!reactGrabImportWithInitMatch) {
    return {
      success: false,
      filePath,
      message: "Could not find React Grab import"
    };
  }
  const optionsJson = formatOptionsAsJson(options);
  const newImport = `import("react-grab").then((m) => m.init(${optionsJson}))`;
  const newContent = originalContent.replace(
    reactGrabImportWithInitMatch[0],
    newImport
  );
  return {
    success: true,
    filePath,
    message: "Update React Grab options",
    originalContent,
    newContent
  };
};
var addOptionsToTanStackImport = (originalContent, options, filePath) => {
  const reactGrabImportWithInitMatch = originalContent.match(
    /(?:void\s+import\s*\(\s*["']react-grab["']\s*\)|import\s*\(\s*["']react-grab\/core["']\s*\)\.then\s*\(\s*\(\s*\{\s*init\s*\}\s*\)\s*=>\s*init\s*\([^)]*\)\s*\))/
  );
  if (!reactGrabImportWithInitMatch) {
    return {
      success: false,
      filePath,
      message: "Could not find React Grab import"
    };
  }
  const optionsJson = formatOptionsAsJson(options);
  const newImport = `import("react-grab/core").then(({ init }) => init(${optionsJson}))`;
  const newContent = originalContent.replace(
    reactGrabImportWithInitMatch[0],
    newImport
  );
  return {
    success: true,
    filePath,
    message: "Update React Grab options",
    originalContent,
    newContent
  };
};
var previewOptionsTransform = (projectRoot, framework, nextRouterType, options) => {
  const filePath = findReactGrabFile(projectRoot, framework, nextRouterType);
  if (!filePath) {
    return {
      success: false,
      filePath: "",
      message: "Could not find file containing React Grab configuration"
    };
  }
  const originalContent = fs.readFileSync(filePath, "utf-8");
  if (!hasReactGrabCode(originalContent)) {
    return {
      success: false,
      filePath,
      message: "Could not find React Grab code in the file"
    };
  }
  switch (framework) {
    case "next":
      return addOptionsToNextScript(originalContent, options, filePath);
    case "vite":
      return addOptionsToViteScript(originalContent, options, filePath);
    case "tanstack":
      return addOptionsToTanStackImport(originalContent, options, filePath);
    case "webpack":
      return addOptionsToWebpackImport(originalContent, options, filePath);
    default:
      return {
        success: false,
        filePath,
        message: `Unknown framework: ${framework}`
      };
  }
};
var applyOptionsTransform = (result) => {
  return applyTransform(result);
};
var removeAgentFromNextApp = (originalContent, agent, filePath) => {
  const agentPackage = `@react-grab/${agent}`;
  if (!originalContent.includes(agentPackage)) {
    return {
      success: true,
      filePath,
      message: `Agent ${agent} is not configured in this file`,
      noChanges: true
    };
  }
  const agentScriptPattern = new RegExp(
    `\\s*\\{process\\.env\\.NODE_ENV === "development" && \\(\\s*<Script[^>]*${agentPackage.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")}[^>]*\\/>\\s*\\)\\}`,
    "gs"
  );
  const simpleScriptPattern = new RegExp(
    `\\s*<Script[^>]*${agentPackage.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")}[^>]*\\/>`,
    "gi"
  );
  let newContent = originalContent.replace(agentScriptPattern, "");
  if (newContent === originalContent) {
    newContent = originalContent.replace(simpleScriptPattern, "");
  }
  if (newContent === originalContent) {
    return {
      success: false,
      filePath,
      message: `Could not find agent ${agent} script to remove`
    };
  }
  return {
    success: true,
    filePath,
    message: `Remove ${agent} agent`,
    originalContent,
    newContent
  };
};
var removeAgentFromVite = (originalContent, agent, filePath) => {
  const agentPackage = `@react-grab/${agent}`;
  if (!originalContent.includes(agentPackage)) {
    return {
      success: true,
      filePath,
      message: `Agent ${agent} is not configured in this file`,
      noChanges: true
    };
  }
  const agentImportPattern = new RegExp(
    `\\s*import\\s*\\(\\s*["']${agentPackage.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")}/client["']\\s*\\);?`,
    "g"
  );
  const newContent = originalContent.replace(agentImportPattern, "");
  if (newContent === originalContent) {
    return {
      success: false,
      filePath,
      message: `Could not find agent ${agent} import to remove`
    };
  }
  return {
    success: true,
    filePath,
    message: `Remove ${agent} agent`,
    originalContent,
    newContent
  };
};
var removeAgentFromWebpack = (originalContent, agent, filePath) => {
  const agentPackage = `@react-grab/${agent}`;
  if (!originalContent.includes(agentPackage)) {
    return {
      success: true,
      filePath,
      message: `Agent ${agent} is not configured in this file`,
      noChanges: true
    };
  }
  const agentImportPattern = new RegExp(
    `\\s*import\\s*\\(\\s*["']${agentPackage.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")}/client["']\\s*\\);?`,
    "g"
  );
  const newContent = originalContent.replace(agentImportPattern, "");
  if (newContent === originalContent) {
    return {
      success: false,
      filePath,
      message: `Could not find agent ${agent} import to remove`
    };
  }
  return {
    success: true,
    filePath,
    message: `Remove ${agent} agent`,
    originalContent,
    newContent
  };
};
var removeAgentFromTanStack = (originalContent, agent, filePath) => {
  const agentPackage = `@react-grab/${agent}`;
  if (!originalContent.includes(agentPackage)) {
    return {
      success: true,
      filePath,
      message: `Agent ${agent} is not configured in this file`,
      noChanges: true
    };
  }
  const agentImportPattern = new RegExp(
    `\\s*void\\s+import\\s*\\(\\s*["']${agentPackage.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")}/client["']\\s*\\);?`,
    "g"
  );
  const newContent = originalContent.replace(agentImportPattern, "");
  if (newContent === originalContent) {
    return {
      success: false,
      filePath,
      message: `Could not find agent ${agent} import to remove`
    };
  }
  return {
    success: true,
    filePath,
    message: `Remove ${agent} agent`,
    originalContent,
    newContent
  };
};
var previewAgentRemoval = (projectRoot, framework, nextRouterType, agent) => {
  const filePath = findReactGrabFile(projectRoot, framework, nextRouterType);
  if (!filePath) {
    return {
      success: true,
      filePath: "",
      message: "Could not find file containing React Grab configuration",
      noChanges: true
    };
  }
  const originalContent = fs.readFileSync(filePath, "utf-8");
  switch (framework) {
    case "next":
      return removeAgentFromNextApp(originalContent, agent, filePath);
    case "vite":
      return removeAgentFromVite(originalContent, agent, filePath);
    case "tanstack":
      return removeAgentFromTanStack(originalContent, agent, filePath);
    case "webpack":
      return removeAgentFromWebpack(originalContent, agent, filePath);
    default:
      return {
        success: false,
        filePath,
        message: `Unknown framework: ${framework}`
      };
  }
};
var previewPackageJsonAgentRemoval = (projectRoot, agent) => {
  const packageJsonPath = path.join(projectRoot, "package.json");
  if (!fs.existsSync(packageJsonPath)) {
    return {
      success: true,
      filePath: "",
      message: "Could not find package.json",
      noChanges: true
    };
  }
  const originalContent = fs.readFileSync(packageJsonPath, "utf-8");
  const allPrefixVariants = getAllAgentPrefixVariants(agent);
  if (allPrefixVariants.length === 0) {
    return {
      success: true,
      filePath: packageJsonPath,
      message: `Unknown agent: ${agent}`,
      noChanges: true
    };
  }
  const hasAnyPrefix = allPrefixVariants.some(
    (prefix) => originalContent.includes(prefix)
  );
  if (!hasAnyPrefix) {
    return {
      success: true,
      filePath: packageJsonPath,
      message: `Agent ${agent} dev script is not configured`,
      noChanges: true
    };
  }
  try {
    const packageJson = JSON.parse(originalContent);
    for (const scriptKey of Object.keys(packageJson.scripts || {})) {
      let scriptValue = packageJson.scripts[scriptKey];
      if (typeof scriptValue === "string") {
        for (const prefix of allPrefixVariants) {
          if (scriptValue.includes(prefix)) {
            scriptValue = scriptValue.replace(prefix + " ", "").replace(prefix, "");
          }
        }
        packageJson.scripts[scriptKey] = scriptValue;
      }
    }
    const newContent = JSON.stringify(packageJson, null, 2) + "\n";
    return {
      success: true,
      filePath: packageJsonPath,
      message: `Remove ${agent} server from dev script`,
      originalContent,
      newContent
    };
  } catch {
    return {
      success: false,
      filePath: packageJsonPath,
      message: "Failed to parse package.json"
    };
  }
};
var previewCdnTransform = (projectRoot, framework, nextRouterType, targetCdnDomain) => {
  const filePath = findReactGrabFile(projectRoot, framework, nextRouterType);
  if (!filePath) {
    return {
      success: false,
      filePath: "",
      message: "Could not find React Grab file"
    };
  }
  const originalContent = fs.readFileSync(filePath, "utf-8");
  const newContent = originalContent.replace(
    /(https?:)?\/\/[^/\s"']+(?=\/(?:@?react-grab))/g,
    `//${targetCdnDomain}`
  ).replace(
    /(https?:)?\/\/[^/\s"']*react-grab[^/\s"']*\.com(?=\/script\.js)/g,
    `//${targetCdnDomain}`
  );
  if (newContent === originalContent) {
    return {
      success: true,
      filePath,
      message: "CDN already set",
      noChanges: true
    };
  }
  return {
    success: true,
    filePath,
    message: "Update CDN",
    originalContent,
    newContent
  };
};

// src/commands/add.ts
var VERSION = "0.1.27";
var formatInstalledAgentNames = (agents) => agents.map((agent) => AGENT_NAMES[agent] || agent).join(", ");
var add = new commander.Command().name("add").alias("install").description("connect React Grab to your agent").argument("[agent]", `agent to connect (${AGENTS.join(", ")}, mcp)`).option("-y, --yes", "skip confirmation prompts", false).option(
  "-c, --cwd <cwd>",
  "working directory (defaults to current directory)",
  process.cwd()
).action(async (agentArg, opts) => {
  console.log(
    `${pc__default.default.magenta("\u273F")} ${pc__default.default.bold("React Grab")} ${pc__default.default.gray(VERSION)}`
  );
  console.log();
  try {
    const cwd = opts.cwd;
    const isNonInteractive = detectNonInteractive(opts.yes);
    const preflightSpinner = spinner("Preflight checks.").start();
    const projectInfo = await detectProject(cwd);
    if (!projectInfo.hasReactGrab) {
      preflightSpinner.fail("React Grab is not installed.");
      logger.break();
      logger.error(
        `Run ${highlighter.info("react-grab init")} first to install React Grab.`
      );
      logger.break();
      process.exit(1);
    }
    preflightSpinner.succeed();
    const availableAgents = AGENTS.filter(
      (agent) => !projectInfo.installedAgents.includes(agent)
    );
    if (availableAgents.length === 0 && isNonInteractive && !agentArg) {
      logger.break();
      logger.success("All legacy agents are already installed.");
      logger.log("Run without -y to add MCP.");
      logger.break();
      process.exit(0);
    }
    let agentIntegration;
    let agentsToRemove = [];
    if (agentArg) {
      if (!AGENTS.includes(agentArg)) {
        logger.break();
        logger.error(`Invalid agent: ${agentArg}`);
        logger.error(`Available agents: ${AGENTS.join(", ")}`);
        logger.break();
        process.exit(1);
      }
      const validAgent = agentArg;
      if (projectInfo.installedAgents.includes(validAgent)) {
        logger.break();
        logger.warn(`${AGENT_NAMES[validAgent]} is already installed.`);
        logger.break();
        process.exit(0);
      }
      agentIntegration = validAgent;
      if (projectInfo.installedAgents.length > 0 && !isNonInteractive) {
        const installedNames = formatInstalledAgentNames(
          projectInfo.installedAgents
        );
        logger.break();
        logger.warn(`${installedNames} is already installed.`);
        const { action } = await prompts({
          type: "select",
          name: "action",
          message: "How would you like to proceed?",
          choices: [
            {
              title: `Replace with ${getAgentDisplayName(agentIntegration)}`,
              value: "replace"
            },
            {
              title: `Add ${getAgentDisplayName(agentIntegration)} alongside existing`,
              value: "add"
            },
            { title: "Cancel", value: "cancel" }
          ]
        });
        if (!action || action === "cancel") {
          logger.break();
          logger.log("Changes cancelled.");
          logger.break();
          process.exit(0);
        }
        if (action === "replace") {
          agentsToRemove = [...projectInfo.installedAgents];
        }
      }
    } else if (!isNonInteractive) {
      if (projectInfo.installedAgents.length > 0) {
        const installedNames = formatInstalledAgentNames(
          projectInfo.installedAgents
        );
        logger.warn(`Currently installed: ${installedNames}`);
        logger.break();
      }
      const connectionMode = await promptConnectionMode();
      if (connectionMode === void 0) {
        logger.break();
        process.exit(1);
      }
      if (connectionMode === "mcp") {
        const didInstall = await promptMcpInstall();
        if (!didInstall) {
          logger.break();
          process.exit(0);
        }
        logger.break();
        logger.log(
          `${highlighter.success("Success!")} MCP server has been configured.`
        );
        logger.log("Restart your agents to activate.");
        logger.break();
        agentIntegration = "mcp";
        projectInfo.installedAgents = [...projectInfo.installedAgents, "mcp"];
      } else {
        const { agent } = await prompts({
          type: "select",
          name: "agent",
          message: `Which ${highlighter.info("agent")} would you like to connect?`,
          choices: [
            ...availableAgents.map((availableAgent) => ({
              title: AGENT_NAMES[availableAgent],
              value: availableAgent
            })),
            { title: "Skip", value: "skip" }
          ]
        });
        if (!agent || agent === "skip") {
          logger.break();
          process.exit(0);
        }
        agentIntegration = agent;
        if (projectInfo.installedAgents.length > 0) {
          const installedNames = formatInstalledAgentNames(
            projectInfo.installedAgents
          );
          const { action } = await prompts({
            type: "select",
            name: "action",
            message: "How would you like to proceed?",
            choices: [
              {
                title: `Replace ${installedNames} with ${getAgentDisplayName(agentIntegration)}`,
                value: "replace"
              },
              {
                title: `Add ${getAgentDisplayName(agentIntegration)} alongside existing`,
                value: "add"
              },
              { title: "Cancel", value: "cancel" }
            ]
          });
          if (!action || action === "cancel") {
            logger.break();
            logger.log("Changes cancelled.");
            logger.break();
            process.exit(0);
          }
          if (action === "replace") {
            agentsToRemove = [...projectInfo.installedAgents];
          }
        }
      }
    } else {
      logger.break();
      logger.error("Please specify an agent to connect.");
      logger.error("Available agents: " + availableAgents.join(", "));
      logger.break();
      process.exit(1);
    }
    if (agentsToRemove.length > 0) {
      for (const agentToRemove of agentsToRemove) {
        const removalResult = previewAgentRemoval(
          projectInfo.projectRoot,
          projectInfo.framework,
          projectInfo.nextRouterType,
          agentToRemove
        );
        const removalPackageJsonResult = previewPackageJsonAgentRemoval(
          projectInfo.projectRoot,
          agentToRemove
        );
        const packagesToRemove = getPackagesToUninstall(agentToRemove);
        if (packagesToRemove.length > 0) {
          const uninstallSpinner = spinner(
            `Removing ${packagesToRemove.join(", ")}.`
          ).start();
          try {
            uninstallPackages(
              packagesToRemove,
              projectInfo.packageManager,
              projectInfo.projectRoot
            );
            uninstallSpinner.succeed();
          } catch (error) {
            uninstallSpinner.fail();
            handleError(error);
          }
        }
        if (removalResult.success && !removalResult.noChanges && removalResult.newContent) {
          const removeWriteSpinner = spinner(
            `Removing ${AGENT_NAMES[agentToRemove]} from ${removalResult.filePath}.`
          ).start();
          const writeResult = applyTransform(removalResult);
          if (!writeResult.success) {
            removeWriteSpinner.fail();
            logger.break();
            logger.error(writeResult.error || "Failed to write file.");
            logger.break();
            process.exit(1);
          }
          removeWriteSpinner.succeed();
        }
        if (removalPackageJsonResult.success && !removalPackageJsonResult.noChanges && removalPackageJsonResult.newContent) {
          const removePackageJsonSpinner = spinner(
            `Removing ${AGENT_NAMES[agentToRemove]} from ${removalPackageJsonResult.filePath}.`
          ).start();
          const packageJsonWriteResult = applyPackageJsonTransform(
            removalPackageJsonResult
          );
          if (!packageJsonWriteResult.success) {
            removePackageJsonSpinner.fail();
            logger.break();
            logger.error(
              packageJsonWriteResult.error || "Failed to write file."
            );
            logger.break();
            process.exit(1);
          }
          removePackageJsonSpinner.succeed();
        }
      }
      projectInfo.installedAgents = projectInfo.installedAgents.filter(
        (installedAgent) => !agentsToRemove.includes(installedAgent)
      );
    }
    const addingSpinner = spinner(
      `Adding ${getAgentDisplayName(agentIntegration)}.`
    ).start();
    addingSpinner.succeed();
    const result = previewTransform(
      projectInfo.projectRoot,
      projectInfo.framework,
      projectInfo.nextRouterType,
      agentIntegration,
      true
    );
    const packageJsonResult = previewPackageJsonTransform(
      projectInfo.projectRoot,
      agentIntegration,
      projectInfo.installedAgents,
      projectInfo.packageManager
    );
    if (!result.success) {
      logger.break();
      logger.error(result.message);
      logger.break();
      process.exit(1);
    }
    const hasLayoutChanges = !result.noChanges && result.originalContent && result.newContent;
    const hasPackageJsonChanges = packageJsonResult.success && !packageJsonResult.noChanges && packageJsonResult.originalContent && packageJsonResult.newContent;
    if (hasLayoutChanges || hasPackageJsonChanges) {
      logger.break();
      if (hasLayoutChanges) {
        printDiff(
          result.filePath,
          result.originalContent,
          result.newContent
        );
      }
      if (hasPackageJsonChanges) {
        if (hasLayoutChanges) {
          logger.break();
        }
        printDiff(
          packageJsonResult.filePath,
          packageJsonResult.originalContent,
          packageJsonResult.newContent
        );
      }
      if (!isNonInteractive && agentsToRemove.length === 0) {
        logger.break();
        const { proceed } = await prompts({
          type: "confirm",
          name: "proceed",
          message: "Apply these changes?",
          initial: true
        });
        if (!proceed) {
          logger.break();
          logger.log("Changes cancelled.");
          logger.break();
          process.exit(0);
        }
      }
    }
    const packages = getPackagesToInstall(agentIntegration, false);
    if (packages.length > 0) {
      const installSpinner = spinner(
        `Installing ${packages.join(", ")}.`
      ).start();
      try {
        installPackages(
          packages,
          projectInfo.packageManager,
          projectInfo.projectRoot
        );
        installSpinner.succeed();
      } catch (error) {
        installSpinner.fail();
        handleError(error);
      }
    }
    if (hasLayoutChanges) {
      const writeSpinner = spinner(
        `Applying changes to ${result.filePath}.`
      ).start();
      const writeResult = applyTransform(result);
      if (!writeResult.success) {
        writeSpinner.fail();
        logger.break();
        logger.error(writeResult.error || "Failed to write file.");
        logger.break();
        process.exit(1);
      }
      writeSpinner.succeed();
    }
    if (hasPackageJsonChanges) {
      const packageJsonSpinner = spinner(
        `Applying changes to ${packageJsonResult.filePath}.`
      ).start();
      const packageJsonWriteResult = applyPackageJsonTransform(packageJsonResult);
      if (!packageJsonWriteResult.success) {
        packageJsonSpinner.fail();
        logger.break();
        logger.error(packageJsonWriteResult.error || "Failed to write file.");
        logger.break();
        process.exit(1);
      }
      packageJsonSpinner.succeed();
    }
    logger.break();
    logger.log(
      `${highlighter.success("Success!")} ${getAgentDisplayName(agentIntegration)} has been added.`
    );
    if (packageJsonResult.warning) {
      logger.warn(packageJsonResult.warning);
    } else {
      logger.log("Make sure to start the agent server before using it.");
    }
    logger.break();
  } catch (error) {
    handleError(error);
  }
});

// src/utils/constants.ts
var MAX_SUGGESTIONS_COUNT = 30;
var MAX_KEY_HOLD_DURATION_MS = 2e3;
var MAX_CONTEXT_LINES = 50;

// src/commands/configure.ts
var VERSION2 = "0.1.27";
var isMac = process.platform === "darwin";
var META_LABEL = isMac ? "Cmd" : "Win";
var ALT_LABEL = isMac ? "Option" : "Alt";
var MODIFIER_ALIASES = {
  cmd: "meta",
  command: "meta",
  win: "meta",
  windows: "meta",
  meta: "meta",
  ctrl: "ctrl",
  control: "ctrl",
  shift: "shift",
  alt: "alt",
  option: "alt",
  opt: "alt"
};
var MODIFIERS = ["meta", "ctrl", "shift", "alt"];
var BASE_KEYS = [
  { key: " ", aliases: ["space", "spacebar"] },
  { key: "Enter", aliases: ["enter", "return"] },
  { key: "Escape", aliases: ["escape", "esc"] },
  { key: "Tab", aliases: ["tab"] },
  { key: "Backspace", aliases: ["backspace", "back"] },
  { key: "Delete", aliases: ["delete", "del"] },
  { key: "Insert", aliases: ["insert", "ins"] },
  { key: "Home", aliases: ["home"] },
  { key: "End", aliases: ["end"] },
  { key: "PageUp", aliases: ["pageup", "pgup"] },
  { key: "PageDown", aliases: ["pagedown", "pgdn", "pgdown"] },
  { key: "ArrowUp", aliases: ["arrowup", "up"] },
  { key: "ArrowDown", aliases: ["arrowdown", "down"] },
  { key: "ArrowLeft", aliases: ["arrowleft", "left"] },
  { key: "ArrowRight", aliases: ["arrowright", "right"] },
  ...Array.from({ length: 12 }, (_, i) => ({
    key: `F${i + 1}`,
    aliases: [`f${i + 1}`]
  })),
  ...Array.from({ length: 26 }, (_, i) => {
    const letter = String.fromCharCode(97 + i);
    return { key: letter, aliases: [letter] };
  }),
  ...Array.from({ length: 10 }, (_, i) => ({
    key: String(i),
    aliases: [String(i)]
  })),
  { key: "`", aliases: ["backtick", "grave", "`"] },
  { key: "-", aliases: ["minus", "dash", "-"] },
  { key: "=", aliases: ["equals", "equal", "="] },
  { key: "[", aliases: ["leftbracket", "lbracket", "["] },
  { key: "]", aliases: ["rightbracket", "rbracket", "]"] },
  { key: "\\", aliases: ["backslash", "\\"] },
  { key: ";", aliases: ["semicolon", ";"] },
  { key: "'", aliases: ["quote", "apostrophe", "'"] },
  { key: ",", aliases: ["comma", ","] },
  { key: ".", aliases: ["period", "dot", "."] },
  { key: "/", aliases: ["slash", "forwardslash", "/"] }
];
var formatCombo = (combo) => {
  const parts = [];
  if (combo.metaKey) parts.push(META_LABEL);
  if (combo.ctrlKey) parts.push("Ctrl");
  if (combo.shiftKey) parts.push("Shift");
  if (combo.altKey) parts.push(ALT_LABEL);
  const keyDisplay = combo.key === " " ? "Space" : combo.key.length === 1 ? combo.key.toUpperCase() : combo.key;
  parts.push(keyDisplay);
  return parts.join("+");
};
var parseInput = (input) => {
  const normalized = input.toLowerCase().replace(/\s+/g, "");
  const parts = normalized.split(/[+\-]/);
  const modifiers = /* @__PURE__ */ new Set();
  let partial = "";
  for (const part of parts) {
    if (!part) continue;
    const modifierKey = MODIFIER_ALIASES[part];
    if (modifierKey) {
      modifiers.add(modifierKey);
    } else {
      partial = part;
    }
  }
  return { modifiers, partial };
};
var POPULAR_KEYS = ["g", "k", "e", "d", "b", " ", "Escape", "Enter"];
var generateSuggestions = (input) => {
  const { modifiers, partial } = parseInput(input);
  const suggestions = [];
  if (!partial && modifiers.size === 0 && !input) {
    for (const mod of MODIFIERS) {
      const label = mod === "meta" ? META_LABEL : mod === "alt" ? ALT_LABEL : mod.charAt(0).toUpperCase() + mod.slice(1);
      for (const popularKey of POPULAR_KEYS) {
        const keyDisplay = popularKey === " " ? "Space" : popularKey.length === 1 ? popularKey.toUpperCase() : popularKey;
        suggestions.push({
          title: `${label}+${keyDisplay}`,
          value: {
            key: popularKey,
            ...mod === "meta" ? { metaKey: true } : {},
            ...mod === "ctrl" ? { ctrlKey: true } : {},
            ...mod === "shift" ? { shiftKey: true } : {},
            ...mod === "alt" ? { altKey: true } : {}
          }
        });
      }
    }
    for (const baseKey of BASE_KEYS) {
      suggestions.push({
        title: baseKey.key === " " ? "Space" : baseKey.key.length === 1 ? baseKey.key.toUpperCase() : baseKey.key,
        value: { key: baseKey.key }
      });
    }
    return suggestions;
  }
  const buildCombo = (key, mods, extraMod) => ({
    key,
    ...mods.has("meta") || extraMod === "meta" ? { metaKey: true } : {},
    ...mods.has("ctrl") || extraMod === "ctrl" ? { ctrlKey: true } : {},
    ...mods.has("shift") || extraMod === "shift" ? { shiftKey: true } : {},
    ...mods.has("alt") || extraMod === "alt" ? { altKey: true } : {}
  });
  for (const baseKey of BASE_KEYS) {
    const matches = partial ? baseKey.aliases.some((alias) => alias.startsWith(partial)) : true;
    if (matches) {
      const combo = buildCombo(baseKey.key, modifiers);
      suggestions.push({
        title: formatCombo(combo),
        value: combo
      });
    }
  }
  if (!partial) {
    const unusedMods = MODIFIERS.filter((m) => !modifiers.has(m));
    for (const mod of unusedMods) {
      for (const popularKey of POPULAR_KEYS) {
        const combo = buildCombo(popularKey, modifiers, mod);
        suggestions.push({
          title: formatCombo(combo),
          value: combo
        });
      }
    }
  }
  return suggestions.slice(0, MAX_SUGGESTIONS_COUNT);
};
var CONFIG_OPTIONS = [
  {
    id: "activationKey",
    title: "Activation Key",
    description: "The key used to activate React Grab (e.g., g, k, space)"
  },
  {
    id: "activationMode",
    title: "Activation Mode",
    description: "Toggle (press to activate/deactivate) or Hold (hold key)"
  },
  {
    id: "keyHoldDuration",
    title: "Key Hold Duration",
    description: "Milliseconds to hold the key before activation (hold mode)"
  },
  {
    id: "allowActivationInsideInput",
    title: "Allow Activation Inside Input",
    description: "Whether to allow activation when focused on input fields"
  },
  {
    id: "maxContextLines",
    title: "Max Context Lines",
    description: "Number of surrounding code lines to include in context"
  }
];
var formatActivationKeyDisplay = (activationKey) => {
  if (!activationKey) return "Default (Option/Alt)";
  return activationKey.split("+").map((part) => {
    const lower = part.toLowerCase();
    if (lower === "meta") return process.platform === "darwin" ? "\u2318" : "Win";
    if (lower === "alt") return process.platform === "darwin" ? "\u2325" : "Alt";
    if (lower === "ctrl") return "Ctrl";
    if (lower === "shift") return "Shift";
    if (lower === "space" || lower === " ") return "Space";
    return part.toUpperCase();
  }).join(" + ");
};
var comboToString = (combo) => {
  const parts = [];
  if (combo.metaKey) parts.push("Meta");
  if (combo.ctrlKey) parts.push("Ctrl");
  if (combo.shiftKey) parts.push("Shift");
  if (combo.altKey) parts.push("Alt");
  if (combo.key) {
    const keyDisplay = combo.key === " " ? "Space" : combo.key;
    parts.push(keyDisplay);
  }
  return parts.join("+");
};
var configure = new commander.Command().name("configure").alias("config").description("configure React Grab options").option("-y, --yes", "skip confirmation prompts", false).option(
  "-k, --key <key>",
  "activation key (e.g., Meta+K, Ctrl+Shift+G, Space)"
).option("-m, --mode <mode>", "activation mode (toggle, hold)").option(
  "--hold-duration <ms>",
  "key hold duration in milliseconds (for hold mode)"
).option(
  "--allow-input <boolean>",
  "allow activation inside input fields (true/false)"
).option("--context-lines <lines>", "max context lines to include").option(
  "--cdn <domain>",
  "CDN domain (e.g., unpkg.com, custom.react-grab.com)"
).option(
  "-c, --cwd <cwd>",
  "working directory (defaults to current directory)",
  process.cwd()
).action(async (opts) => {
  console.log(
    `${pc__default.default.magenta("\u273F")} ${pc__default.default.bold("React Grab")} ${pc__default.default.gray(VERSION2)}`
  );
  console.log();
  try {
    const cwd = opts.cwd;
    const preflightSpinner = spinner("Preflight checks.").start();
    const projectInfo = await detectProject(cwd);
    if (!projectInfo.hasReactGrab) {
      preflightSpinner.fail("React Grab is not installed.");
      logger.break();
      logger.error(
        `Run ${highlighter.info("react-grab init")} first to install React Grab.`
      );
      logger.break();
      process.exit(1);
    }
    preflightSpinner.succeed();
    if (opts.cdn) {
      const result2 = previewCdnTransform(
        projectInfo.projectRoot,
        projectInfo.framework,
        projectInfo.nextRouterType,
        opts.cdn
      );
      if (!result2.success) {
        logger.break();
        logger.error(result2.message);
        logger.break();
        process.exit(1);
      }
      if (result2.noChanges) {
        logger.break();
        logger.log("No changes needed.");
        logger.break();
        process.exit(0);
      }
      logger.break();
      printDiff(result2.filePath, result2.originalContent, result2.newContent);
      if (!opts.yes) {
        logger.break();
        const { proceed } = await prompts({
          type: "confirm",
          name: "proceed",
          message: "Apply these changes?",
          initial: true
        });
        if (!proceed) {
          logger.break();
          logger.log("Changes cancelled.");
          logger.break();
          process.exit(0);
        }
      }
      const writeSpinner = spinner(
        `Applying changes to ${result2.filePath}.`
      ).start();
      const writeResult = applyTransform(result2);
      if (!writeResult.success) {
        writeSpinner.fail();
        logger.break();
        logger.error(writeResult.error || "Failed to write file.");
        logger.break();
        process.exit(1);
      }
      writeSpinner.succeed();
      logger.break();
      logger.log(`${highlighter.success("Success!")} CDN updated.`);
      logger.break();
      return;
    }
    const hasFlags = opts.key || opts.mode || opts.holdDuration || opts.allowInput || opts.contextLines;
    logger.break();
    logger.log(`Configure ${highlighter.info("React Grab")} options:`);
    logger.break();
    const collectedOptions = {};
    if (hasFlags) {
      if (opts.key) {
        collectedOptions.activationKey = opts.key;
        logger.log(
          `  Activation key: ${highlighter.info(formatActivationKeyDisplay(collectedOptions.activationKey))}`
        );
      }
      if (opts.mode) {
        if (opts.mode !== "toggle" && opts.mode !== "hold") {
          logger.error(`Invalid mode: ${opts.mode}. Use "toggle" or "hold".`);
          logger.break();
          process.exit(1);
        }
        collectedOptions.activationMode = opts.mode;
        logger.log(`  Activation mode: ${highlighter.info(opts.mode)}`);
      }
      if (opts.holdDuration) {
        const duration = parseInt(opts.holdDuration, 10);
        if (isNaN(duration) || duration < 0 || duration > MAX_KEY_HOLD_DURATION_MS) {
          logger.error(
            `Invalid hold duration. Must be 0-${MAX_KEY_HOLD_DURATION_MS}ms.`
          );
          logger.break();
          process.exit(1);
        }
        collectedOptions.keyHoldDuration = duration;
        logger.log(
          `  Key hold duration: ${highlighter.info(`${duration}ms`)}`
        );
      }
      if (opts.allowInput !== void 0) {
        const allowInput = opts.allowInput === "true" || opts.allowInput === true;
        collectedOptions.allowActivationInsideInput = allowInput;
        logger.log(
          `  Allow activation inside input: ${highlighter.info(String(allowInput))}`
        );
      }
      if (opts.contextLines) {
        const lines = parseInt(opts.contextLines, 10);
        if (isNaN(lines) || lines < 0 || lines > MAX_CONTEXT_LINES) {
          logger.error(
            `Invalid context lines. Must be 0-${MAX_CONTEXT_LINES}.`
          );
          logger.break();
          process.exit(1);
        }
        collectedOptions.maxContextLines = lines;
        logger.log(`  Max context lines: ${highlighter.info(String(lines))}`);
      }
    } else {
      const { selectedOption } = await prompts({
        type: "autocomplete",
        name: "selectedOption",
        message: "Search for an option to configure:",
        choices: CONFIG_OPTIONS.map((option) => ({
          title: option.title,
          value: option.id,
          description: option.description
        })),
        suggest: (input, choices) => Promise.resolve(
          choices.filter(
            (choice) => choice.title.toLowerCase().includes(input.toLowerCase()) || (choice.description?.toLowerCase().includes(input.toLowerCase()) ?? false)
          )
        )
      });
      if (selectedOption === void 0) {
        logger.break();
        process.exit(1);
      }
      if (selectedOption === "activationKey") {
        const { selectedCombo } = await prompts({
          type: "autocomplete",
          name: "selectedCombo",
          message: "Type key combination (e.g. ctrl+shift+g):",
          choices: generateSuggestions(""),
          suggest: (input) => Promise.resolve(generateSuggestions(input))
        });
        if (selectedCombo === void 0) {
          logger.break();
          process.exit(1);
        }
        collectedOptions.activationKey = comboToString(selectedCombo);
        logger.log(
          `  Activation key: ${highlighter.info(formatActivationKeyDisplay(collectedOptions.activationKey))}`
        );
      }
      if (selectedOption === "activationMode") {
        const { activationMode } = await prompts({
          type: "select",
          name: "activationMode",
          message: `Select ${highlighter.info("activation mode")}:`,
          choices: [
            {
              title: "Toggle (press to activate/deactivate)",
              value: "toggle"
            },
            { title: "Hold (hold key to keep active)", value: "hold" }
          ],
          initial: 0
        });
        if (activationMode === void 0) {
          logger.break();
          process.exit(1);
        }
        collectedOptions.activationMode = activationMode;
      }
      if (selectedOption === "keyHoldDuration") {
        const { keyHoldDuration } = await prompts({
          type: "number",
          name: "keyHoldDuration",
          message: `Enter ${highlighter.info("key hold duration")} in milliseconds:`,
          initial: 150,
          min: 0,
          max: 2e3
        });
        if (keyHoldDuration === void 0) {
          logger.break();
          process.exit(1);
        }
        collectedOptions.keyHoldDuration = keyHoldDuration;
      }
      if (selectedOption === "allowActivationInsideInput") {
        const { allowActivationInsideInput } = await prompts({
          type: "confirm",
          name: "allowActivationInsideInput",
          message: `Allow activation ${highlighter.info("inside input fields")}?`,
          initial: true
        });
        if (allowActivationInsideInput === void 0) {
          logger.break();
          process.exit(1);
        }
        collectedOptions.allowActivationInsideInput = allowActivationInsideInput;
      }
      if (selectedOption === "maxContextLines") {
        const { maxContextLines } = await prompts({
          type: "number",
          name: "maxContextLines",
          message: `Enter ${highlighter.info("max context lines")} to include:`,
          initial: 3,
          min: 0,
          max: 50
        });
        if (maxContextLines === void 0) {
          logger.break();
          process.exit(1);
        }
        collectedOptions.maxContextLines = maxContextLines;
      }
    }
    const result = previewOptionsTransform(
      projectInfo.projectRoot,
      projectInfo.framework,
      projectInfo.nextRouterType,
      collectedOptions
    );
    if (!result.success) {
      logger.break();
      logger.warn(result.message);
      logger.break();
      const configJson = JSON.stringify(collectedOptions);
      logger.log(
        `Add this to your ${highlighter.info("init()")} call or ${highlighter.info("data-options")} attribute:`
      );
      logger.break();
      console.log(`  ${pc__default.default.cyan(configJson)}`);
      logger.break();
      process.exit(1);
    }
    const hasChanges = !result.noChanges && result.originalContent && result.newContent;
    if (hasChanges) {
      logger.break();
      printDiff(result.filePath, result.originalContent, result.newContent);
      if (!opts.yes) {
        logger.break();
        const { proceed } = await prompts({
          type: "confirm",
          name: "proceed",
          message: "Apply these changes?",
          initial: true
        });
        if (!proceed) {
          logger.break();
          logger.log("Changes cancelled.");
          logger.break();
          process.exit(0);
        }
      }
      const writeSpinner = spinner(
        `Applying changes to ${result.filePath}.`
      ).start();
      const writeResult = applyOptionsTransform(result);
      if (!writeResult.success) {
        writeSpinner.fail();
        logger.break();
        logger.error(writeResult.error || "Failed to write file.");
        logger.break();
        process.exit(1);
      }
      writeSpinner.succeed();
    } else {
      logger.break();
      logger.log("No changes needed.");
    }
    logger.break();
    logger.log(
      `${highlighter.success("Success!")} React Grab options have been configured.`
    );
    logger.break();
  } catch (error) {
    handleError(error);
  }
});

// src/utils/cli-helpers.ts
var formatInstalledAgentNames2 = (agents) => agents.map(getAgentDisplayName).join(", ");
var applyTransformWithFeedback = (result, message) => {
  const writeSpinner = spinner(
    message ?? `Applying changes to ${result.filePath}.`
  ).start();
  const writeResult = applyTransform(result);
  if (!writeResult.success) {
    writeSpinner.fail();
    logger.break();
    logger.error(writeResult.error || "Failed to write file.");
    logger.break();
    process.exit(1);
  }
  writeSpinner.succeed();
};
var applyPackageJsonWithFeedback = (result, message) => {
  const writeSpinner = spinner(
    message ?? `Applying changes to ${result.filePath}.`
  ).start();
  const writeResult = applyPackageJsonTransform(result);
  if (!writeResult.success) {
    writeSpinner.fail();
    logger.break();
    logger.error(writeResult.error || "Failed to write file.");
    logger.break();
    process.exit(1);
  }
  writeSpinner.succeed();
};
var installPackagesWithFeedback = (packages, packageManager, projectRoot) => {
  if (packages.length === 0) return;
  const installSpinner = spinner(`Installing ${packages.join(", ")}.`).start();
  try {
    installPackages(packages, packageManager, projectRoot);
    installSpinner.succeed();
  } catch (error) {
    installSpinner.fail();
    handleError(error);
  }
};
var uninstallPackagesWithFeedback = (packages, packageManager, projectRoot) => {
  if (packages.length === 0) return;
  const uninstallSpinner = spinner(`Removing ${packages.join(", ")}.`).start();
  try {
    uninstallPackages(packages, packageManager, projectRoot);
    uninstallSpinner.succeed();
  } catch (error) {
    uninstallSpinner.fail();
    handleError(error);
  }
};

// src/commands/init.ts
var VERSION3 = "0.1.27";
var REPORT_URL = "https://react-grab.com/api/report-cli";
var DOCS_URL = "https://github.com/aidenybai/react-grab";
var reportToCli = (type, config, error) => {
  fetch(REPORT_URL, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      type,
      version: VERSION3,
      config,
      error: error ? { message: error.message, stack: error.stack } : void 0,
      timestamp: (/* @__PURE__ */ new Date()).toISOString()
    })
  }).catch(() => {
  });
};
var FRAMEWORK_NAMES = {
  next: "Next.js",
  vite: "Vite",
  tanstack: "TanStack Start",
  webpack: "Webpack",
  unknown: "Unknown"
};
var PACKAGE_MANAGER_NAMES = {
  npm: "npm",
  yarn: "Yarn",
  pnpm: "pnpm",
  bun: "Bun"
};
var UNSUPPORTED_FRAMEWORK_NAMES = {
  remix: "Remix",
  astro: "Astro",
  sveltekit: "SvelteKit",
  gatsby: "Gatsby"
};
var getAgentName = getAgentDisplayName;
var sortProjectsByFramework = (projects) => [...projects].sort((projectA, projectB) => {
  if (projectA.framework === "unknown" && projectB.framework !== "unknown")
    return 1;
  if (projectA.framework !== "unknown" && projectB.framework === "unknown")
    return -1;
  return 0;
});
var printSubprojects = (searchRoot, sortedProjects) => {
  logger.break();
  logger.log("Found the following projects:");
  logger.break();
  for (const project of sortedProjects) {
    const frameworkLabel = project.framework !== "unknown" ? ` ${highlighter.dim(`(${FRAMEWORK_NAMES[project.framework]})`)}` : "";
    const relativePath = path.relative(searchRoot, project.path);
    logger.log(
      `  ${highlighter.info(project.name)}${frameworkLabel} ${highlighter.dim(relativePath)}`
    );
  }
  logger.break();
  logger.log(
    `Re-run with ${highlighter.info("-c <path>")} to specify a project:`
  );
  logger.break();
  logger.log(
    `  ${highlighter.dim("$")} npx -y grab@latest init -c ${path.relative(searchRoot, sortedProjects[0].path)}`
  );
  logger.break();
};
var formatActivationKeyDisplay2 = (activationKey) => {
  if (!activationKey) return "Default (Option/Alt)";
  return activationKey.split("+").map((part) => {
    const lower = part.toLowerCase();
    if (lower === "meta") return process.platform === "darwin" ? "\u2318" : "Win";
    if (lower === "alt") return process.platform === "darwin" ? "\u2325" : "Alt";
    if (lower === "ctrl") return "Ctrl";
    if (lower === "shift") return "Shift";
    if (lower === "space" || lower === " ") return "Space";
    return part.toUpperCase();
  }).join(" + ");
};
var init = new commander.Command().name("init").description("initialize React Grab in your project").option("-y, --yes", "skip confirmation prompts", false).option("-f, --force", "force overwrite existing config", false).option(
  "-a, --agent <agent>",
  `connect to your agent (${AGENTS.join(", ")}, mcp)`
).option(
  "-k, --key <key>",
  "activation key (e.g., Meta+K, Ctrl+Shift+G, Space)"
).option("--skip-install", "skip package installation", false).option("--pkg <pkg>", "custom package URL for CLI (e.g., grab)").option(
  "-c, --cwd <cwd>",
  "working directory (defaults to current directory)",
  process.cwd()
).action(async (opts) => {
  console.log(
    `${pc__default.default.magenta("\u273F")} ${pc__default.default.bold("React Grab")} ${pc__default.default.gray(VERSION3)}`
  );
  console.log();
  try {
    const cwd = path.resolve(opts.cwd);
    const isNonInteractive = detectNonInteractive(opts.yes);
    if (!fs.existsSync(cwd)) {
      logger.break();
      logger.error(`Directory does not exist: ${highlighter.info(cwd)}`);
      logger.break();
      process.exit(1);
    }
    const preflightSpinner = spinner("Preflight checks.").start();
    const projectInfo = await detectProject(cwd);
    const removeAgents = async (agentsToRemove2, skipInstall = false) => {
      for (const agentToRemove of agentsToRemove2) {
        const removalResult = previewAgentRemoval(
          projectInfo.projectRoot,
          projectInfo.framework,
          projectInfo.nextRouterType,
          agentToRemove
        );
        const removalPackageJsonResult = previewPackageJsonAgentRemoval(
          projectInfo.projectRoot,
          agentToRemove
        );
        if (!skipInstall) {
          uninstallPackagesWithFeedback(
            getPackagesToUninstall(agentToRemove),
            projectInfo.packageManager,
            projectInfo.projectRoot
          );
        }
        if (removalResult.success && !removalResult.noChanges && removalResult.newContent) {
          applyTransformWithFeedback(
            removalResult,
            `Removing ${getAgentName(agentToRemove)} from ${removalResult.filePath}.`
          );
        }
        if (removalPackageJsonResult.success && !removalPackageJsonResult.noChanges && removalPackageJsonResult.newContent) {
          applyPackageJsonWithFeedback(
            removalPackageJsonResult,
            `Removing ${getAgentName(agentToRemove)} from ${removalPackageJsonResult.filePath}.`
          );
        }
      }
    };
    if (projectInfo.hasReactGrab && !opts.force) {
      preflightSpinner.succeed();
      if (isNonInteractive) {
        logger.break();
        logger.warn("React Grab is already installed.");
        logger.log(
          `Use ${highlighter.info("--force")} to reconfigure, or remove ${highlighter.info("--yes")} for interactive mode.`
        );
        logger.break();
        process.exit(0);
      }
      logger.break();
      logger.success("React Grab is already installed.");
      logger.break();
      if (projectInfo.installedAgents.length > 0) {
        logger.log(
          `Currently installed agents: ${highlighter.info(formatInstalledAgentNames2(projectInfo.installedAgents))}`
        );
        logger.break();
      }
      const { wantCustomizeOptions } = await prompts({
        type: "confirm",
        name: "wantCustomizeOptions",
        message: `Would you like to customize ${highlighter.info("options")}?`,
        initial: false
      });
      if (wantCustomizeOptions === void 0) {
        logger.break();
        process.exit(1);
      }
      if (wantCustomizeOptions || opts.key) {
        logger.break();
        logger.log(`Configure ${highlighter.info("React Grab")} options:`);
        logger.break();
        const collectedOptions = {};
        if (opts.key) {
          collectedOptions.activationKey = opts.key;
          logger.log(
            `  Activation key: ${highlighter.info(formatActivationKeyDisplay2(collectedOptions.activationKey))}`
          );
        } else {
          const { wantActivationKey } = await prompts({
            type: "confirm",
            name: "wantActivationKey",
            message: `Configure ${highlighter.info("activation key")}?`,
            initial: false
          });
          if (wantActivationKey === void 0) {
            logger.break();
            process.exit(1);
          }
          if (wantActivationKey) {
            const { key } = await prompts({
              type: "text",
              name: "key",
              message: "Enter the activation key (e.g., g, k, space):",
              initial: ""
            });
            if (key === void 0) {
              logger.break();
              process.exit(1);
            }
            collectedOptions.activationKey = key ? key.toLowerCase() : void 0;
            logger.log(
              `  Activation key: ${highlighter.info(formatActivationKeyDisplay2(collectedOptions.activationKey))}`
            );
          }
        }
        const { activationMode } = await prompts({
          type: "select",
          name: "activationMode",
          message: `Select ${highlighter.info("activation mode")}:`,
          choices: [
            {
              title: "Toggle (press to activate/deactivate)",
              value: "toggle"
            },
            { title: "Hold (hold key to keep active)", value: "hold" }
          ],
          initial: 0
        });
        if (activationMode === void 0) {
          logger.break();
          process.exit(1);
        }
        collectedOptions.activationMode = activationMode;
        if (activationMode === "hold") {
          const { keyHoldDuration } = await prompts({
            type: "number",
            name: "keyHoldDuration",
            message: `Enter ${highlighter.info("key hold duration")} in milliseconds:`,
            initial: 150,
            min: 0,
            max: 2e3
          });
          if (keyHoldDuration === void 0) {
            logger.break();
            process.exit(1);
          }
          collectedOptions.keyHoldDuration = keyHoldDuration;
        }
        const { allowActivationInsideInput } = await prompts({
          type: "confirm",
          name: "allowActivationInsideInput",
          message: `Allow activation ${highlighter.info("inside input fields")}?`,
          initial: true
        });
        if (allowActivationInsideInput === void 0) {
          logger.break();
          process.exit(1);
        }
        collectedOptions.allowActivationInsideInput = allowActivationInsideInput;
        const { maxContextLines } = await prompts({
          type: "number",
          name: "maxContextLines",
          message: `Enter ${highlighter.info("max context lines")} to include:`,
          initial: 3,
          min: 0,
          max: 50
        });
        if (maxContextLines === void 0) {
          logger.break();
          process.exit(1);
        }
        collectedOptions.maxContextLines = maxContextLines;
        const optionsResult = previewOptionsTransform(
          projectInfo.projectRoot,
          projectInfo.framework,
          projectInfo.nextRouterType,
          collectedOptions
        );
        if (!optionsResult.success) {
          logger.break();
          logger.error(optionsResult.message);
          logger.break();
          process.exit(1);
        }
        const hasOptionsChanges = !optionsResult.noChanges && optionsResult.originalContent && optionsResult.newContent;
        if (hasOptionsChanges) {
          logger.break();
          printDiff(
            optionsResult.filePath,
            optionsResult.originalContent,
            optionsResult.newContent
          );
          logger.break();
          const { proceed } = await prompts({
            type: "confirm",
            name: "proceed",
            message: "Apply these changes?",
            initial: true
          });
          if (!proceed) {
            logger.break();
            logger.log("Options configuration cancelled.");
          } else {
            applyTransformWithFeedback(optionsResult);
            logger.break();
            logger.success("React Grab options have been configured.");
          }
        } else {
          logger.break();
          logger.log("No option changes needed.");
        }
      }
      const availableAgents = AGENTS.filter(
        (agent) => !projectInfo.installedAgents.includes(agent)
      );
      logger.break();
      const { wantAddAgent } = await prompts({
        type: "confirm",
        name: "wantAddAgent",
        message: `Would you like to ${highlighter.info("connect it to your agent")}?`,
        initial: false
      });
      if (wantAddAgent === void 0) {
        logger.break();
        process.exit(1);
      }
      if (wantAddAgent) {
        const connectionMode = await promptConnectionMode();
        if (connectionMode === void 0) {
          logger.break();
          process.exit(1);
        }
        let agentIntegration2;
        if (connectionMode === "mcp") {
          const didInstall = await promptMcpInstall();
          if (!didInstall) {
            logger.break();
            process.exit(0);
          }
          logger.break();
          logger.success("MCP server has been configured.");
          logger.log("Restart your agents to activate.");
          agentIntegration2 = "mcp";
          projectInfo.installedAgents = ["mcp"];
          const result2 = previewTransform(
            projectInfo.projectRoot,
            projectInfo.framework,
            projectInfo.nextRouterType,
            agentIntegration2,
            true
          );
          const packageJsonResult2 = previewPackageJsonTransform(
            projectInfo.projectRoot,
            agentIntegration2,
            projectInfo.installedAgents,
            projectInfo.packageManager
          );
          if (!result2.success) {
            logger.break();
            logger.error(result2.message);
            logger.break();
            process.exit(1);
          }
          const hasLayoutChanges2 = !result2.noChanges && result2.originalContent && result2.newContent;
          const hasPackageJsonChanges2 = packageJsonResult2.success && !packageJsonResult2.noChanges && packageJsonResult2.originalContent && packageJsonResult2.newContent;
          if (hasLayoutChanges2 || hasPackageJsonChanges2) {
            logger.break();
            if (hasLayoutChanges2) {
              printDiff(
                result2.filePath,
                result2.originalContent,
                result2.newContent
              );
            }
            if (hasPackageJsonChanges2) {
              if (hasLayoutChanges2) {
                logger.break();
              }
              printDiff(
                packageJsonResult2.filePath,
                packageJsonResult2.originalContent,
                packageJsonResult2.newContent
              );
            }
            logger.break();
            const { proceed } = await prompts({
              type: "confirm",
              name: "proceed",
              message: "Apply these changes?",
              initial: true
            });
            if (!proceed) {
              logger.break();
              logger.log("Agent addition cancelled.");
            } else {
              installPackagesWithFeedback(
                getPackagesToInstall(agentIntegration2, false),
                projectInfo.packageManager,
                projectInfo.projectRoot
              );
              if (hasLayoutChanges2) {
                applyTransformWithFeedback(result2);
              }
              if (hasPackageJsonChanges2) {
                applyPackageJsonWithFeedback(packageJsonResult2);
              }
              logger.break();
              logger.success(
                `${getAgentName(agentIntegration2)} has been added.`
              );
            }
          }
        } else {
          const { agent } = await prompts({
            type: "select",
            name: "agent",
            message: `Which ${highlighter.info("agent")} would you like to connect?`,
            choices: [
              ...availableAgents.map((innerAgent) => ({
                title: getAgentName(innerAgent),
                value: innerAgent
              })),
              { title: "Skip", value: "skip" }
            ]
          });
          if (agent === void 0 || agent === "skip") {
            logger.break();
            process.exit(0);
          }
          agentIntegration2 = agent;
          let agentsToRemove2 = [];
          if (projectInfo.installedAgents.length > 0) {
            const installedNames = formatInstalledAgentNames2(
              projectInfo.installedAgents
            );
            const { action } = await prompts({
              type: "select",
              name: "action",
              message: "How would you like to proceed?",
              choices: [
                {
                  title: `Replace ${installedNames} with ${getAgentName(agentIntegration2)}`,
                  value: "replace"
                },
                {
                  title: `Add ${getAgentName(agentIntegration2)} alongside existing`,
                  value: "add"
                },
                { title: "Cancel", value: "cancel" }
              ]
            });
            if (!action || action === "cancel") {
              logger.break();
              logger.log("Agent addition cancelled.");
            } else {
              if (action === "replace") {
                agentsToRemove2 = [...projectInfo.installedAgents];
              }
              if (agentsToRemove2.length > 0) {
                await removeAgents(agentsToRemove2);
                projectInfo.installedAgents = projectInfo.installedAgents.filter(
                  (innerAgent) => !agentsToRemove2.includes(innerAgent)
                );
              }
              const result2 = previewTransform(
                projectInfo.projectRoot,
                projectInfo.framework,
                projectInfo.nextRouterType,
                agentIntegration2,
                true
              );
              const packageJsonResult2 = previewPackageJsonTransform(
                projectInfo.projectRoot,
                agentIntegration2,
                projectInfo.installedAgents,
                projectInfo.packageManager
              );
              if (!result2.success) {
                logger.break();
                logger.error(result2.message);
                logger.break();
                process.exit(1);
              }
              const hasLayoutChanges2 = !result2.noChanges && result2.originalContent && result2.newContent;
              const hasPackageJsonChanges2 = packageJsonResult2.success && !packageJsonResult2.noChanges && packageJsonResult2.originalContent && packageJsonResult2.newContent;
              if (hasLayoutChanges2 || hasPackageJsonChanges2) {
                logger.break();
                if (hasLayoutChanges2) {
                  printDiff(
                    result2.filePath,
                    result2.originalContent,
                    result2.newContent
                  );
                }
                if (hasPackageJsonChanges2) {
                  if (hasLayoutChanges2) {
                    logger.break();
                  }
                  printDiff(
                    packageJsonResult2.filePath,
                    packageJsonResult2.originalContent,
                    packageJsonResult2.newContent
                  );
                }
                if (agentsToRemove2.length === 0) {
                  logger.break();
                  const { proceed } = await prompts({
                    type: "confirm",
                    name: "proceed",
                    message: "Apply these changes?",
                    initial: true
                  });
                  if (!proceed) {
                    logger.break();
                    logger.log("Agent addition cancelled.");
                  } else {
                    installPackagesWithFeedback(
                      getPackagesToInstall(agentIntegration2, false),
                      projectInfo.packageManager,
                      projectInfo.projectRoot
                    );
                    if (hasLayoutChanges2) {
                      applyTransformWithFeedback(result2);
                    }
                    if (hasPackageJsonChanges2) {
                      applyPackageJsonWithFeedback(packageJsonResult2);
                    }
                    logger.break();
                    logger.success(
                      `${getAgentName(agentIntegration2)} has been added.`
                    );
                  }
                } else {
                  installPackagesWithFeedback(
                    getPackagesToInstall(agentIntegration2, false),
                    projectInfo.packageManager,
                    projectInfo.projectRoot
                  );
                  if (hasLayoutChanges2) {
                    applyTransformWithFeedback(result2);
                  }
                  if (hasPackageJsonChanges2) {
                    applyPackageJsonWithFeedback(packageJsonResult2);
                  }
                  logger.break();
                  logger.success(
                    `${getAgentName(agentIntegration2)} has been added.`
                  );
                }
              }
            }
          } else {
            const result2 = previewTransform(
              projectInfo.projectRoot,
              projectInfo.framework,
              projectInfo.nextRouterType,
              agentIntegration2,
              true
            );
            const packageJsonResult2 = previewPackageJsonTransform(
              projectInfo.projectRoot,
              agentIntegration2,
              projectInfo.installedAgents,
              projectInfo.packageManager
            );
            if (!result2.success) {
              logger.break();
              logger.error(result2.message);
              logger.break();
              process.exit(1);
            }
            const hasLayoutChanges2 = !result2.noChanges && result2.originalContent && result2.newContent;
            const hasPackageJsonChanges2 = packageJsonResult2.success && !packageJsonResult2.noChanges && packageJsonResult2.originalContent && packageJsonResult2.newContent;
            if (hasLayoutChanges2 || hasPackageJsonChanges2) {
              logger.break();
              if (hasLayoutChanges2) {
                printDiff(
                  result2.filePath,
                  result2.originalContent,
                  result2.newContent
                );
              }
              if (hasPackageJsonChanges2) {
                if (hasLayoutChanges2) {
                  logger.break();
                }
                printDiff(
                  packageJsonResult2.filePath,
                  packageJsonResult2.originalContent,
                  packageJsonResult2.newContent
                );
              }
              logger.break();
              const { proceed } = await prompts({
                type: "confirm",
                name: "proceed",
                message: "Apply these changes?",
                initial: true
              });
              if (!proceed) {
                logger.break();
                logger.log("Agent addition cancelled.");
              } else {
                installPackagesWithFeedback(
                  getPackagesToInstall(agentIntegration2, false),
                  projectInfo.packageManager,
                  projectInfo.projectRoot
                );
                if (hasLayoutChanges2) {
                  applyTransformWithFeedback(result2);
                }
                if (hasPackageJsonChanges2) {
                  applyPackageJsonWithFeedback(packageJsonResult2);
                }
                logger.break();
                logger.success(
                  `${getAgentName(agentIntegration2)} has been added.`
                );
              }
            }
          }
        }
      }
      logger.break();
      process.exit(0);
    }
    preflightSpinner.succeed();
    const frameworkSpinner = spinner("Verifying framework.").start();
    if (projectInfo.unsupportedFramework) {
      const frameworkName = UNSUPPORTED_FRAMEWORK_NAMES[projectInfo.unsupportedFramework];
      frameworkSpinner.fail(`Found ${highlighter.info(frameworkName)}.`);
      logger.break();
      logger.log(`${frameworkName} is not yet supported by automatic setup.`);
      logger.log(`Visit ${highlighter.info(DOCS_URL)} for manual setup.`);
      logger.break();
      process.exit(1);
    }
    if (projectInfo.framework === "unknown") {
      let searchRoot = cwd;
      let reactProjects = findReactProjects(searchRoot);
      if (reactProjects.length === 0 && cwd !== process.cwd()) {
        searchRoot = process.cwd();
        reactProjects = findReactProjects(searchRoot);
      }
      if (reactProjects.length > 0) {
        frameworkSpinner.info(
          `Verifying framework. Found ${reactProjects.length} project${reactProjects.length === 1 ? "" : "s"}.`
        );
        const sortedProjects = sortProjectsByFramework(reactProjects);
        if (isNonInteractive) {
          printSubprojects(searchRoot, sortedProjects);
          process.exit(1);
        }
        logger.break();
        const { selectedProject } = await prompts({
          type: "select",
          name: "selectedProject",
          message: "Select a project to install React Grab:",
          choices: [
            ...sortedProjects.map((project) => {
              const frameworkLabel = project.framework !== "unknown" ? ` ${highlighter.dim(`(${FRAMEWORK_NAMES[project.framework]})`)}` : "";
              return {
                title: `${project.name}${frameworkLabel}`,
                value: project.path
              };
            }),
            { title: "Skip", value: "skip" }
          ]
        });
        if (!selectedProject || selectedProject === "skip") {
          logger.break();
          process.exit(0);
        }
        process.chdir(selectedProject);
        const newProjectInfo = await detectProject(selectedProject);
        Object.assign(projectInfo, newProjectInfo);
        const newFrameworkSpinner = spinner("Verifying framework.").start();
        newFrameworkSpinner.succeed(
          `Verifying framework. Found ${highlighter.info(FRAMEWORK_NAMES[newProjectInfo.framework])}.`
        );
      } else {
        frameworkSpinner.fail("Could not detect a supported framework.");
        logger.break();
        logger.log(
          "React Grab supports Next.js, Vite, TanStack Start, and Webpack projects."
        );
        logger.log(`Visit ${highlighter.info(DOCS_URL)} for manual setup.`);
        logger.break();
        process.exit(1);
      }
    } else {
      frameworkSpinner.succeed(
        `Verifying framework. Found ${highlighter.info(FRAMEWORK_NAMES[projectInfo.framework])}.`
      );
    }
    if (projectInfo.framework === "next") {
      const routerSpinner = spinner("Detecting router type.").start();
      routerSpinner.succeed(
        `Detecting router type. Found ${highlighter.info(projectInfo.nextRouterType === "app" ? "App Router" : "Pages Router")}.`
      );
    }
    const packageManagerSpinner = spinner(
      "Detecting package manager."
    ).start();
    packageManagerSpinner.succeed(
      `Detecting package manager. Found ${highlighter.info(PACKAGE_MANAGER_NAMES[projectInfo.packageManager])}.`
    );
    const finalFramework = projectInfo.framework;
    const finalPackageManager = projectInfo.packageManager;
    const finalNextRouterType = projectInfo.nextRouterType;
    let agentIntegration = opts.agent || "none";
    const agentsToRemove = [];
    if (!isNonInteractive && !opts.agent) {
      logger.break();
      const { wantAddAgent } = await prompts({
        type: "confirm",
        name: "wantAddAgent",
        message: `Would you like to ${highlighter.info("connect it to your agent")}?`,
        initial: false
      });
      if (wantAddAgent === void 0) {
        logger.break();
        process.exit(1);
      }
      if (wantAddAgent) {
        const connectionMode = await promptConnectionMode();
        if (connectionMode === void 0) {
          logger.break();
          process.exit(1);
        }
        if (connectionMode === "mcp") {
          const didInstall = await promptMcpInstall();
          if (!didInstall) {
            logger.break();
            process.exit(0);
          }
          logger.break();
          logger.success("MCP server has been configured.");
          logger.log("Continuing with React Grab installation...");
          logger.break();
          agentIntegration = "mcp";
        } else {
          const { agent } = await prompts({
            type: "select",
            name: "agent",
            message: `Which ${highlighter.info("agent")} would you like to connect?`,
            choices: [
              ...AGENTS.map((innerAgent) => ({
                title: getAgentName(innerAgent),
                value: innerAgent
              })),
              { title: "Skip", value: "skip" }
            ]
          });
          if (agent === void 0) {
            logger.break();
            process.exit(1);
          }
          if (agent !== "skip") {
            agentIntegration = agent;
          }
        }
      }
    }
    const result = previewTransform(
      projectInfo.projectRoot,
      finalFramework,
      finalNextRouterType,
      agentIntegration,
      false,
      opts.force
    );
    const packageJsonResult = previewPackageJsonTransform(
      projectInfo.projectRoot,
      agentIntegration,
      projectInfo.installedAgents,
      finalPackageManager
    );
    if (!result.success) {
      logger.break();
      logger.error(result.message);
      logger.error(`Visit ${highlighter.info(DOCS_URL)} for manual setup.`);
      logger.break();
      process.exit(1);
    }
    const hasLayoutChanges = !result.noChanges && result.originalContent && result.newContent;
    const hasPackageJsonChanges = packageJsonResult.success && !packageJsonResult.noChanges && packageJsonResult.originalContent && packageJsonResult.newContent;
    if (hasLayoutChanges || hasPackageJsonChanges) {
      logger.break();
      if (hasLayoutChanges) {
        printDiff(
          result.filePath,
          result.originalContent,
          result.newContent
        );
      }
      if (hasPackageJsonChanges) {
        if (hasLayoutChanges) {
          logger.break();
        }
        printDiff(
          packageJsonResult.filePath,
          packageJsonResult.originalContent,
          packageJsonResult.newContent
        );
      }
      logger.break();
      logger.warn("Auto-detection may not be 100% accurate.");
      logger.warn("Please verify the changes before committing.");
      if (!isNonInteractive) {
        logger.break();
        const { proceed } = await prompts({
          type: "confirm",
          name: "proceed",
          message: "Apply these changes?",
          initial: true
        });
        if (!proceed) {
          logger.break();
          logger.log("Changes cancelled.");
          logger.break();
          process.exit(0);
        }
      }
    }
    if (agentsToRemove.length > 0) {
      await removeAgents(agentsToRemove, opts.skipInstall);
      projectInfo.installedAgents = projectInfo.installedAgents.filter(
        (agent) => !agentsToRemove.includes(agent)
      );
    }
    const shouldInstallReactGrab = !projectInfo.hasReactGrab;
    const shouldInstallAgent = agentIntegration !== "none" && !projectInfo.installedAgents.includes(agentIntegration);
    if (!opts.skipInstall && (shouldInstallReactGrab || shouldInstallAgent)) {
      installPackagesWithFeedback(
        getPackagesToInstall(agentIntegration, shouldInstallReactGrab),
        finalPackageManager,
        projectInfo.projectRoot
      );
    }
    if (hasLayoutChanges) {
      applyTransformWithFeedback(result);
    }
    if (hasPackageJsonChanges) {
      applyPackageJsonWithFeedback(packageJsonResult);
    }
    logger.break();
    logger.log(
      `${highlighter.success("Success!")} React Grab has been installed.`
    );
    if (packageJsonResult.warning) {
      logger.break();
      logger.warn(packageJsonResult.warning);
      logger.break();
    } else {
      logger.log("You may now start your development server.");
    }
    logger.break();
    reportToCli("completed", {
      framework: finalFramework,
      packageManager: finalPackageManager,
      router: finalNextRouterType,
      agent: agentIntegration !== "none" ? agentIntegration : void 0,
      isMonorepo: projectInfo.isMonorepo
    });
  } catch (error) {
    handleError(error);
    reportToCli("error", void 0, error);
  }
});
var VERSION4 = "0.1.27";
var remove = new commander.Command().name("remove").description("disconnect React Grab from your agent").argument("[agent]", `agent to disconnect (${AGENTS.join(", ")}, mcp)`).option("-y, --yes", "skip confirmation prompts", false).option(
  "-c, --cwd <cwd>",
  "working directory (defaults to current directory)",
  process.cwd()
).action(async (agentArg, opts) => {
  console.log(
    `${pc__default.default.magenta("\u273F")} ${pc__default.default.bold("React Grab")} ${pc__default.default.gray(VERSION4)}`
  );
  console.log();
  try {
    const cwd = opts.cwd;
    const isNonInteractive = detectNonInteractive(opts.yes);
    const preflightSpinner = spinner("Preflight checks.").start();
    const projectInfo = await detectProject(cwd);
    if (!projectInfo.hasReactGrab) {
      preflightSpinner.fail("React Grab is not installed.");
      logger.break();
      logger.error(
        `Run ${highlighter.info("react-grab init")} first to install React Grab.`
      );
      logger.break();
      process.exit(1);
    }
    if (projectInfo.installedAgents.length === 0) {
      preflightSpinner.succeed();
      logger.break();
      logger.warn("No agent connections are installed.");
      logger.break();
      process.exit(0);
    }
    preflightSpinner.succeed();
    let agentToRemove;
    if (agentArg) {
      if (!projectInfo.installedAgents.includes(agentArg)) {
        logger.break();
        logger.error(`Agent ${highlighter.info(agentArg)} is not installed.`);
        logger.log(
          `Installed agents: ${projectInfo.installedAgents.map(getAgentDisplayName).join(", ")}`
        );
        logger.break();
        process.exit(1);
      }
      agentToRemove = agentArg;
    } else if (!isNonInteractive) {
      logger.break();
      const { agent } = await prompts({
        type: "select",
        name: "agent",
        message: `Which ${highlighter.info("agent")} would you like to disconnect?`,
        choices: projectInfo.installedAgents.map((innerAgent) => ({
          title: getAgentDisplayName(innerAgent),
          value: innerAgent
        }))
      });
      if (!agent) {
        logger.break();
        process.exit(1);
      }
      agentToRemove = agent;
    } else {
      logger.break();
      logger.error("Please specify an agent to disconnect.");
      logger.error(
        "Installed agents: " + projectInfo.installedAgents.join(", ")
      );
      logger.break();
      process.exit(1);
    }
    const removingSpinner = spinner(
      `Preparing to remove ${getAgentDisplayName(agentToRemove)}.`
    ).start();
    removingSpinner.succeed();
    const result = previewAgentRemoval(
      projectInfo.projectRoot,
      projectInfo.framework,
      projectInfo.nextRouterType,
      agentToRemove
    );
    const packageJsonResult = previewPackageJsonAgentRemoval(
      projectInfo.projectRoot,
      agentToRemove
    );
    const hasLayoutChanges = result.success && !result.noChanges && result.originalContent && result.newContent;
    const hasPackageJsonChanges = packageJsonResult.success && !packageJsonResult.noChanges && packageJsonResult.originalContent && packageJsonResult.newContent;
    if (hasLayoutChanges || hasPackageJsonChanges) {
      logger.break();
      if (hasLayoutChanges) {
        printDiff(
          result.filePath,
          result.originalContent,
          result.newContent
        );
      }
      if (hasPackageJsonChanges) {
        if (hasLayoutChanges) {
          logger.break();
        }
        printDiff(
          packageJsonResult.filePath,
          packageJsonResult.originalContent,
          packageJsonResult.newContent
        );
      }
      if (!isNonInteractive) {
        logger.break();
        const { proceed } = await prompts({
          type: "confirm",
          name: "proceed",
          message: "Apply these changes?",
          initial: true
        });
        if (!proceed) {
          logger.break();
          logger.log("Changes cancelled.");
          logger.break();
          process.exit(0);
        }
      }
    }
    const packages = getPackagesToUninstall(agentToRemove);
    if (packages.length > 0) {
      const uninstallSpinner = spinner(
        `Removing ${packages.join(", ")}.`
      ).start();
      try {
        uninstallPackages(
          packages,
          projectInfo.packageManager,
          projectInfo.projectRoot
        );
        uninstallSpinner.succeed();
      } catch (error) {
        uninstallSpinner.fail();
        handleError(error);
      }
    }
    if (hasLayoutChanges) {
      const writeSpinner = spinner(
        `Applying changes to ${result.filePath}.`
      ).start();
      const writeResult = applyTransform(result);
      if (!writeResult.success) {
        writeSpinner.fail();
        logger.break();
        logger.error(writeResult.error || "Failed to write file.");
        logger.break();
        process.exit(1);
      }
      writeSpinner.succeed();
    }
    if (hasPackageJsonChanges) {
      const packageJsonSpinner = spinner(
        `Applying changes to ${packageJsonResult.filePath}.`
      ).start();
      const packageJsonWriteResult = applyPackageJsonTransform(packageJsonResult);
      if (!packageJsonWriteResult.success) {
        packageJsonSpinner.fail();
        logger.break();
        logger.error(packageJsonWriteResult.error || "Failed to write file.");
        logger.break();
        process.exit(1);
      }
      packageJsonSpinner.succeed();
    }
    logger.break();
    logger.log(
      `${highlighter.success("Success!")} ${getAgentDisplayName(agentToRemove)} has been removed.`
    );
    logger.break();
  } catch (error) {
    handleError(error);
  }
});

// src/cli.ts
var VERSION5 = "0.1.27";
var VERSION_API_URL = "https://www.react-grab.com/api/version";
process.on("SIGINT", () => process.exit(0));
process.on("SIGTERM", () => process.exit(0));
try {
  fetch(`${VERSION_API_URL}?source=cli&t=${Date.now()}`).catch(() => {
  });
} catch {
}
var program = new commander.Command().name("grab").description("add React Grab to your project").version(VERSION5, "-v, --version", "display the version number");
program.addCommand(init);
program.addCommand(add);
program.addCommand(remove);
program.addCommand(configure);
var main = async () => {
  await program.parseAsync();
};
main();
